<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Load data</title>

<script src="site_libs/header-attrs-2.7/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-sm-12 col-md-4 col-lg-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-sm-12 col-md-8 col-lg-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">STUtility</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="About.html">About</a>
</li>
<li>
  <a href="Installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Getting_Started.html">Load data</a>
    </li>
    <li>
      <a href="Subset_and_Merge.html">Subset and Merge</a>
    </li>
    <li>
      <a href="Quality_Control.html">Quality Control</a>
    </li>
    <li>
      <a href="Plots_and_Themes.html">Plots and Themes</a>
    </li>
    <li>
      <a href="Image_Processing.html">Image processing</a>
    </li>
    <li>
      <a href="Manual_Annotation.html">Manual annotation</a>
    </li>
    <li>
      <a href="Normalization.html">Normalization</a>
    </li>
    <li>
      <a href="Spatial_Features.html">Spatial features</a>
    </li>
    <li>
      <a href="Neighborhood.html">Neighborhood Analysis</a>
    </li>
    <li>
      <a href="3D_visualization.html">3D visualization</a>
    </li>
    <li>
      <a href="Split_Data.html">Split Data</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Walkthroughs
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Maize_data.html">Maize data analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="License.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ludvigla/STUtility_web_site">
    <span class="fas fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Load data</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-02-28
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>STUtility_web_site/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version 1.7.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20191031code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20191031)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20191031code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20191031)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomludviglaSTUtilitywebsitetree64ae8bec1e7c302f54af1fd298a32004f122294atargetblank64ae8bea"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/ludvigla/STUtility_web_site/tree/64ae8bec1e7c302f54af1fd298a32004f122294a" target="_blank">64ae8be</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomludviglaSTUtilitywebsitetree64ae8bec1e7c302f54af1fd298a32004f122294atargetblank64ae8bea" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/ludvigla/STUtility_web_site/tree/64ae8bec1e7c302f54af1fd298a32004f122294a" target="_blank">64ae8be</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/manual_annotation.png
    Ignored:    pre_data/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/Getting_Started.Rmd</code>) and HTML (<code>docs/Getting_Started.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/d41bcb071d3d8ac3ff52f854616aaceebf61dcb6/docs/Getting_Started.html" target="_blank">d41bcb0</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2022-02-28
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/0dafcee92125a64ba11fd8e833fbd65d63393dfd/docs/Getting_Started.html" target="_blank">0dafcee</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2021-05-06
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/df62517bf936c9743f2f47db57c91163b3c5c153/docs/Getting_Started.html" target="_blank">df62517</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2021-05-06
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/88b046fb086e401566789f8510f234382b3ecca3/docs/Getting_Started.html" target="_blank">88b046f</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2021-05-05
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/2523b7054e8831a69210b53fefc78404634e1086/docs/Getting_Started.html" target="_blank">2523b70</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2021-05-05
</td>
<td>
Changed names of docs
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ludvigla/STUtility_web_site/blob/363f9112fa844fc3d1ec14dfda1c491f8ef8e749/analysis/Getting_Started.Rmd" target="_blank">363f911</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2021-05-05
</td>
<td>
Changed names of Rmds
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<style type="text/css">
div.main-container {
  background-color: #FFFFFF !important;
  max-width: 1800px;
  margin-left: auto;
  margin-right: auto;
}
</style>
<style>
#TOC {
  background: url("https://www.spatialresearch.org/wp-content/uploads/2019/09/str-logo-spatial_research_3@2x.png");
  background-size: contain;
  padding-top: 100px !important;
  background-repeat: no-repeat;
  width: 500px;
  color: black;
  border-color: #FFFFFF !important;
}
</style>
<style> code, pre{
  background-color: #F8F8F8 !important;
  color: black !important;
}
</style>
<style>
.watch-out { background-color: #F0F0F0; border: 1px #F0F0F0; }
<style>
<style> 
body {
  color: black;
  font-size: 20pt;
}
</style>
<p>First you need to load the library into your R session.</p>
<pre class="r"><code>library(STutility)</code></pre>
<div id="x-visium-platform" class="section level2">
<h2>10X Visium platform</h2>
<hr />
<div id="input-files" class="section level3">
<h3>Input files</h3>
<hr />
<p>10X Visium data output is produced with the <a href="https://support.10xgenomics.com/spatial-gene-expression/software/visualization/latest/installation">spaceranger</a> command line tool from raw fastq files. The output includes a number of files, and the ones that needs to be imported into R for STutility are the following:</p>
<ol style="list-style-type: decimal">
<li>“filtered_feature_bc_matrix.h5” or “raw_feature_bc_matrix.h5” : Gene expression matrices in .h5 format containing the raw UMI counts for each spot and each gene. The “filtered_feature_bc_matrix.h5” matrix contains spots that are localized under the tissue while the “raw_feature_bc_matrix.h5” matrix contain all spots from the entire capture area. NOTE: if you want to include spots outside of the tissue, you need to set <code>disable.subset = TRUE</code> when running <code>InputFromTable</code>.</li>
<li>“tissue_positions_list.csv” : contains capture-spot barcode and capture-spot coordinates.</li>
<li>“tissue_hires_image.png” : H&amp;E image in PNG format with a width of 2000 pixels.</li>
<li>“scalefactors_json.json” : This file contains scaling factors subject to the H&amp;E images of different resolutions. E.g. “tissue_hires_scalef”: 0.1, means that the pixel coordinates in the “tissue_positions_list.csv” table should be scaled by a factor of 0.1 to match the size of the “hires_image.png” file.</li>
</ol>
<p>To use the full range of functions within STutility, all four files are needed for each sample. However, all data analysis steps that do not involve the H&amp;E image can be performed with only the count file as input. To read in the 10x Visium .h5 files, the package <code>hdf5r</code> needs to be installed: <code>BiocManager::install("hdf5r")</code>.</p>
<hr />
<p>To follow along this tutorial with a test data set, go to the <a href="https://support.10xgenomics.com/spatial-gene-expression/datasets">10x Dataset repo</a> and download the following two files:</p>
<ul>
<li>Feature / cell matrix HDF5 (filtered)</li>
<li>Spatial imaging data (.zip)
<ul>
<li>tissue_hires_image</li>
<li>tissue_positions_list</li>
<li>scalefactors_json</li>
</ul></li>
</ul>
<p>The .zip file contains the H&amp;E image (in two resolution formats; “tissue_lowres_image” and “tissue_hires_image”), the “tissue_positions_list” with pixel coordinates for the orginial TIF image and the “scalefactors_json.json” that contains the scalefactors used to derive the pixel cooridinates for the hires images. There are three alternatives to handle the scaling of pixel coordinates. Either, you manualy open the JSON file and note the scalefactor and state these numbers in a column in the infoTable named “scaleVisium” (see below). Or, you add a column named “json” with paths to the “scalefactors_json.json” files. A third option is to manually input the values to the function InputFromTable (see <code>?InputFromTable</code>). If the scalefactors are incorrect, you will end up with misaligned coordinates. This could for example happen if you are trying to use the orignal H&amp;E image and not the “tissue_hires_image.png” when running <code>InputFromTable</code>.</p>
<p>In this vignette we have used the datasets from Mouse Brain Serial Section 1 and 2 (Sagittal-Posterior)</p>
</div>
<div id="prepare-data" class="section level3">
<h3>Prepare data</h3>
<p>The recommended method to read the files into R is via the creation of a <code>data.frame</code> that we will call the <code>infoTable</code>. There are four columns in this table that are required for Visium data: “samples”, “spotfiles”, “imgs” and “json”. These columns specify the paths to the required input files.</p>
<br>
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:800px; ">
<table class=" lightable-paper" style="font-family: &quot;Arial Narrow&quot;, arial, helvetica, sans-serif; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
samples
</th>
<th style="text-align:left;">
spotfiles
</th>
<th style="text-align:left;">
imgs
</th>
<th style="text-align:left;">
json
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
path/to/sample_1/count_file_1.h5
</td>
<td style="text-align:left;">
path/to/sample_1/tissue_positions_list.csv
</td>
<td style="text-align:left;">
path/to/sample_1/tissue_hires_image.png
</td>
<td style="text-align:left;">
path/to/sample_1/scalefactors_json.json
</td>
</tr>
<tr>
<td style="text-align:left;">
path/to/sample_2/count_file_2.h5
</td>
<td style="text-align:left;">
path/to/sample_2/tissue_positions_list.csv
</td>
<td style="text-align:left;">
path/to/sample_2/tissue_hires_image.png
</td>
<td style="text-align:left;">
path/to/sample_2/scalefactors_json.json
</td>
</tr>
</tbody>
</table>
</div>
<p>Any number of <em>extra</em> columns can be added to the infoTable data.frame that you want to include as meta data in your Seurat object, e.g. “gender”, “age”, “slide_id” etc. These columns can be named as you like, but they should not be called <em>not</em> “sample”, “spotfiles”, “imgs” or “json”.</p>
<p>We are now ready to load our samples and create a “Seurat” object using our infotTable <code>data.frame</code>.</p>
<p>Here, we demonstrate the creation of the Seurat object, while also including some filtering (see section “Quality Control” for more information on filtering):</p>
<ul>
<li>Keeping the genes that are found in at least 5 capture spots and has a total count value &gt;= 100.</li>
<li>Keeping the capture-spots that contains &gt;= 500 total transcripts.</li>
</ul>
<p>Note that you have to specify which platform the data comes from. The default platform is 10X Visium but if you wish to run data from the older ST platforms, there is support for “1k” and “2k” arrays. You can also mix datasets from different platforms by specifying one of; “Visium”, “1k” or “2k” in a separate column of the infoTable named “platform”. You just have to make sure that the datasets have gene symbols which follows the same nomenclature so that count matrices can be merged. <br></p>
<pre class="r"><code>se &lt;- InputFromTable(infotable = infoTable, 
                      min.gene.count = 100, 
                      min.gene.spots = 5,
                      min.spot.count = 500,
                      platform =  &quot;Visium&quot;)</code></pre>
<p><br> Once you have created a Seurat object you can process and visualize your data just like in a scRNA-seq experiment and make use of the plethora of functions provided in the <code>Seurat</code> package. There are many vignettes to get started available at the <a href="https://satijalab.org/seurat/vignettes.html">Seurat web site</a>.</p>
<p>For example, if you wish to explore the spatial distribution of various features on the array coordinates you can do this using the <code>ST.FeaturePlot()</code> function. Features include any column stored in the “meta.data” slot, dimensionality reduction objects or gene expression vectors.</p>
<p><br></p>
<pre class="r"><code>ST.FeaturePlot(se, features = c(&quot;nFeature_RNA&quot;), cols = c(&quot;lightgray&quot;, &quot;mistyrose&quot;, &quot;red&quot;, &quot;darkred&quot;, &quot;black&quot;), ncol = 2, pt.size = 1.3)</code></pre>
<p><img src="figure/Getting_Started.Rmd/unnamed-chunk-1-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-1-1">
Past versions of unnamed-chunk-1-1.png
</button>
</p>
<div id="fig-unnamed-chunk-1-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ludvigla/STUtility_web_site/blob/a2036a86d4e3f9192caa4dff8df7ee38ac632001/docs/figure/Getting_Started.Rmd/unnamed-chunk-1-1.png" target="_blank">a2036a8</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2021-05-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><br></p>
</div>
</div>
<div id="older-st-platforms" class="section level2">
<h2>Older ST platforms</h2>
<p>In general, using STutility for the old ST platform data follows the same workflow as for the 10X Visium arrays. The only difference is when loading the data into R.</p>
<div id="input-files-1" class="section level3">
<h3>Input files</h3>
<p>The original ST workflow produces the following three output files:</p>
<ol style="list-style-type: decimal">
<li>Count file in TSV format (Count file with raw counts (UMI filtered) for each gene and capture spot)</li>
<li>Spot detector output (File with spatial pixel coordinate information produced using the <a href="https://github.com/SpatialTranscriptomicsResearch/st_spot_detector">Spot Detector webtool</a>)</li>
<li>H&amp;E image (same as the one used for spot detection)</li>
</ol>
</div>
<div id="prepare-data-1" class="section level3">
<h3>Prepare data</h3>
<p>The recommended method to read the files into R is via the creation of a “infoTable”, which is a table with at least three columns “samples”, “spotfiles” and “imgs”.</p>
<p>Test data:</p>
<pre class="r"><code>samples &lt;- list.files(pattern = &quot;^Hippo[1-6].tsv.gz&quot;, path = system.file(&quot;extdata/counts&quot;, package = &quot;STutility&quot;), full.names = TRUE)
spotfiles &lt;- list.files(pattern = &quot;^alignment_table_Hippo[1-6].tsv.gz&quot;, path = system.file(&quot;extdata/counts&quot;, package = &quot;STutility&quot;), full.names = TRUE)
imgs &lt;- list.files(pattern = &quot;^Hippo[1-6].jpg&quot;, path = system.file(&quot;extdata/counts&quot;, package = &quot;STutility&quot;), full.names = TRUE)

infoTable &lt;- data.frame(samples, imgs, spotfiles, stringsAsFactors = F)</code></pre>
</div>
<div id="load-data-and-convert-from-ensambleids-to-gene-symbols" class="section level3">
<h3>Load data and convert from EnsambleIDs to gene symbols</h3>
<p>The provided count matrices uses EnsambleIDs (with version id) for the gene symbols. Gene symbols are often a preference for easier reading, and we have therefore included an option to directly convert the gene IDs when creating the Seurat object. The data.frame object required for conversion should have one column called “gene_id” matching the original gene IDs and a column called “gene_name” with the desired symbols. you also need to make sure that these columns have unique names, otherwise the converiion will not work. We have provided such a table that you can use to convert between EnsambleIDs and MGI symbols (mus musculus gene nomenclature).</p>
<p><br></p>
<pre class="r"><code>#Transformation table for geneIDs
ensids &lt;- read.table(file = list.files(system.file(&quot;extdata&quot;, package = &quot;STutility&quot;), full.names = T, pattern = &quot;mouse_genes&quot;), header = T, sep = &quot;\t&quot;, stringsAsFactors = F)</code></pre>
<p>We are now ready to load our samples and create a “Seurat” object.</p>
<p>Here, we demonstrate the creation of the Seurat object, while also including some filtering:</p>
<ul>
<li>Keeping the genes that are found in at least 5 capture spots and has a total count value &gt;= 100.</li>
<li>Keeping the capture-spots that contains &gt;= 500 total transcripts.</li>
</ul>
<p>Note that we specify that we’re using the “2k” array platform and also, since we in this case have genes in the columns, we set transpose=TRUE. <br></p>
<pre class="r"><code>#TODO: add warnings if ids missmatch. Check that ids are in the data.frame ...
se &lt;- InputFromTable(infotable = infoTable, 
                      min.gene.count = 100, 
                      min.gene.spots = 5,
                      min.spot.count = 500, 
                      annotation = ensids, 
                      platform = &quot;2k&quot;,
                      transpose = TRUE)</code></pre>
<p><br> Once you have created a Seurat object you can process and visualize your data just like in a scRNA-seq experiment and make use of the plethora of functions provided in the Seurat package. There are many vignettes to get started available at the <a href="https://satijalab.org/seurat/vignettes.html">Seurat web site</a>.</p>
<p>Some of the functionalities provided in the Seurat package are not yet supported by STutility, such as dataset integration and multimodal analysis. These methods should in principle work if you treat the data like a scRNA-seq experiment, but you will not be able to make use of the image related data or the spatial visualization functions.</p>
<p>For example, if you wish to explore the spatial distribution of various features on the array coordinates you can do this using the <code>ST.FeaturePlot()</code> function. <br></p>
<pre class="r"><code>ST.FeaturePlot(se, features = c(&quot;nFeature_RNA&quot;), cols = c(&quot;lightgray&quot;, &quot;mistyrose&quot;, &quot;red&quot;, &quot;darkred&quot;, &quot;black&quot;), ncol = 2, pt.size = 1.5, show.sb = FALSE)</code></pre>
<p><img src="figure/Getting_Started.Rmd/pot_fts-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-pot_fts-1">
Past versions of pot_fts-1.png
</button>
</p>
<div id="fig-pot_fts-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ludvigla/STUtility_web_site/blob/a2036a86d4e3f9192caa4dff8df7ee38ac632001/docs/figure/Getting_Started.Rmd/pot_fts-1.png" target="_blank">a2036a8</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2021-05-05
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><br></p>
</div>
</div>
<div id="navigating-and-accessing-data" class="section level2">
<h2>Navigating and accessing data</h2>
<hr />
<p>We recommend that unexperienced users have look at the <code>Seurat</code> website and tutorials for basic navigation of the Seurat object such as getting and setting identities and accessing various method outputs.</p>
<p>However, specific for STutility, there is another S4 class object stored inside the <code>Seurat</code> object, called <code>Staffli</code>. This object contains all the STutility specific meta data, like pixel cooridinates, sample IDs, platform types etc.</p>
<p>You can reach this via:</p>
<pre class="r"><code>st.object &lt;- GetStaffli(se)
st.object</code></pre>
<pre><code>An object of class Staffli 
4557 spots across 4 samples. 

Available image representations: 
   raw, masked, masked.masks, processed, processed.masks </code></pre>
<pre class="r"><code># Access meta data
head(st.object[[]]) %&gt;%
  kbl() %&gt;%
  kable_styling()</code></pre>
<table class="table" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
x
</th>
<th style="text-align:right;">
y
</th>
<th style="text-align:right;">
adj_x
</th>
<th style="text-align:right;">
adj_y
</th>
<th style="text-align:right;">
pixel_x
</th>
<th style="text-align:right;">
pixel_y
</th>
<th style="text-align:left;">
sample
</th>
<th style="text-align:right;">
warped_x
</th>
<th style="text-align:right;">
warped_y
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
2x20_1
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
20
</td>
<td style="text-align:right;">
2.01
</td>
<td style="text-align:right;">
19.99
</td>
<td style="text-align:right;">
1431.9
</td>
<td style="text-align:right;">
2640.9
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:right;">
1431.9
</td>
<td style="text-align:right;">
2640.9
</td>
</tr>
<tr>
<td style="text-align:left;">
2x22_1
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
22
</td>
<td style="text-align:right;">
2.01
</td>
<td style="text-align:right;">
21.98
</td>
<td style="text-align:right;">
1432.2
</td>
<td style="text-align:right;">
2880.2
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:right;">
1432.2
</td>
<td style="text-align:right;">
2880.2
</td>
</tr>
<tr>
<td style="text-align:left;">
2x24_1
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
24
</td>
<td style="text-align:right;">
2.01
</td>
<td style="text-align:right;">
23.99
</td>
<td style="text-align:right;">
1432.0
</td>
<td style="text-align:right;">
3122.0
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:right;">
1432.0
</td>
<td style="text-align:right;">
3122.0
</td>
</tr>
<tr>
<td style="text-align:left;">
2x26_1
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
26
</td>
<td style="text-align:right;">
2.03
</td>
<td style="text-align:right;">
26.00
</td>
<td style="text-align:right;">
1434.0
</td>
<td style="text-align:right;">
3363.4
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:right;">
1434.0
</td>
<td style="text-align:right;">
3363.4
</td>
</tr>
<tr>
<td style="text-align:left;">
2x28_1
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
28
</td>
<td style="text-align:right;">
1.98
</td>
<td style="text-align:right;">
27.97
</td>
<td style="text-align:right;">
1429.3
</td>
<td style="text-align:right;">
3600.9
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:right;">
1429.3
</td>
<td style="text-align:right;">
3600.9
</td>
</tr>
<tr>
<td style="text-align:left;">
2x30_1
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
2.00
</td>
<td style="text-align:right;">
29.95
</td>
<td style="text-align:right;">
1431.3
</td>
<td style="text-align:right;">
3839.2
</td>
<td style="text-align:left;">
1
</td>
<td style="text-align:right;">
1431.3
</td>
<td style="text-align:right;">
3839.2
</td>
</tr>
</tbody>
</table>
<p>It can for example be useful to access the spot coordinates and images if you want to write your own plotting functions.</p>
<div id="important-note" class="section level3">
<h3>Important note</h3>
<hr />
<p>Because the spatial infromation is stored in a slot that is normally not part of the <code>Seurat</code> environment, some function such as <code>subset</code> or <code>merge</code> will not work as expected. Instead, you should use the <code>SubsetSTData</code> and <code>MergeSTData</code> functions from STutility.</p>
 
<hr />
<p style="text-align: center;">
A work by <a href="mailto: j.bergenstrahle@scilifelab.se">Joseph Bergenstråhle</a> and <a href="mailto: ludvig.larsson@scilifelab.se">Ludvig Larsson</a>
</p>
<p><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css"></p>
<p style="text-align: center;">
<a href="https://www.spatialresearch.org" class="fa fa-beer"></a>
</p>
<p> </p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.3 (2020-10-10)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS/LAPACK: /Users/ludviglarsson/anaconda3/envs/R4.0/lib/libopenblasp-r0.3.12.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] magrittr_2.0.1     kableExtra_1.3.4   STutility_0.1.0    ggplot2_3.3.5     
[5] SeuratObject_4.0.0 Seurat_4.0.2       workflowr_1.7.0   

loaded via a namespace (and not attached):
  [1] utf8_1.2.1              reticulate_1.18         tidyselect_1.1.1       
  [4] htmlwidgets_1.5.3       grid_4.0.3              Rtsne_0.15             
  [7] munsell_0.5.0           codetools_0.2-18        ica_1.0-2              
 [10] units_0.7-1             future_1.21.0           miniUI_0.1.1.1         
 [13] withr_2.4.1             colorspace_2.0-0        highr_0.8              
 [16] knitr_1.31              uuid_0.1-4              rstudioapi_0.13        
 [19] ROCR_1.0-11             tensor_1.5              listenv_0.8.0          
 [22] labeling_0.4.2          git2r_0.28.0            polyclip_1.10-0        
 [25] farver_2.1.0            rprojroot_2.0.2         coda_0.19-4            
 [28] parallelly_1.25.0       LearnBayes_2.15.1       vctrs_0.3.8            
 [31] generics_0.1.0          xfun_0.20               R6_2.5.0               
 [34] doParallel_1.0.16       Morpho_2.8              ggiraph_0.7.8          
 [37] manipulateWidget_0.11.0 spatstat.utils_2.2-0    assertthat_0.2.1       
 [40] promises_1.2.0.1        scales_1.1.1            imager_0.42.8          
 [43] gtable_0.3.0            globals_0.14.0          bmp_0.3                
 [46] processx_3.5.1          goftest_1.2-2           rlang_1.0.1            
 [49] zeallot_0.1.0           akima_0.6-2.1           systemfonts_1.0.1      
 [52] splines_4.0.3           lazyeval_0.2.2          spatstat.geom_2.3-0    
 [55] rgl_0.105.22            yaml_2.2.1              reshape2_1.4.4         
 [58] abind_1.4-5             crosstalk_1.1.1         httpuv_1.5.5           
 [61] tools_4.0.3             spData_0.3.8            ellipsis_0.3.2         
 [64] spatstat.core_2.3-0     raster_3.4-10           jquerylib_0.1.3        
 [67] RColorBrewer_1.1-2      proxy_0.4-25            Rvcg_0.19.2            
 [70] ggridges_0.5.3          Rcpp_1.0.6              plyr_1.8.6             
 [73] classInt_0.4-3          purrr_0.3.4             ps_1.6.0               
 [76] rpart_4.1-15            dbscan_1.1-6            deldir_1.0-6           
 [79] pbapply_1.4-3           viridis_0.6.1           cowplot_1.1.1          
 [82] zoo_1.8-9               ggrepel_0.9.1           cluster_2.1.1          
 [85] colorRamps_2.3          fs_1.5.0                data.table_1.14.0      
 [88] magick_2.7.2            scattermore_0.7         readbitmap_0.1.5       
 [91] gmodels_2.18.1          lmtest_0.9-38           RANN_2.6.1             
 [94] whisker_0.4             fitdistrplus_1.1-3      matrixStats_0.58.0     
 [97] patchwork_1.1.1         shinyjs_2.0.0           mime_0.10              
[100] evaluate_0.14           xtable_1.8-4            jpeg_0.1-8.1           
[103] gridExtra_2.3           compiler_4.0.3          tibble_3.1.6           
[106] KernSmooth_2.23-18      crayon_1.4.1            htmltools_0.5.1.1      
[109] mgcv_1.8-34             later_1.1.0.1           spdep_1.1-7            
[112] tiff_0.1-8              tidyr_1.2.0             expm_0.999-6           
[115] DBI_1.1.1               MASS_7.3-53.1           sf_0.9-8               
[118] boot_1.3-27             Matrix_1.3-2            cli_3.1.1              
[121] gdata_2.18.0            parallel_4.0.3          igraph_1.2.6           
[124] pkgconfig_2.0.3         getPass_0.2-2           sp_1.4-5               
[127] plotly_4.9.3            spatstat.sparse_2.0-0   xml2_1.3.2             
[130] foreach_1.5.1           svglite_2.0.0           bslib_0.2.4            
[133] webshot_0.5.2           rvest_1.0.0             stringr_1.4.0          
[136] callr_3.7.0             digest_0.6.27           sctransform_0.3.2      
[139] RcppAnnoy_0.0.18        spatstat.data_2.1-0     rmarkdown_2.7          
[142] leiden_0.3.7            uwot_0.1.10             gdtools_0.2.3          
[145] shiny_1.6.0             gtools_3.8.2            lifecycle_1.0.1        
[148] nlme_3.1-152            jsonlite_1.7.2          viridisLite_0.4.0      
[151] fansi_0.4.2             pillar_1.7.0            lattice_0.20-41        
[154] fastmap_1.1.0           httr_1.4.2              survival_3.2-10        
[157] glue_1.4.2              png_0.1-7               iterators_1.0.13       
[160] class_7.3-18            stringi_1.5.3           sass_0.3.1             
[163] dplyr_1.0.8             irlba_2.3.3             e1071_1.7-6            
[166] future.apply_1.7.0     </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
