<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="" />


<title>Image processing</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">STUtility</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Index.html">Home</a>
</li>
<li>
  <a href="About.html">About</a>
</li>
<li>
  <a href="Installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Getting started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Getting_Started.html">Load data</a>
    </li>
    <li>
      <a href="Quality_Control.html">Quality Control</a>
    </li>
    <li>
      <a href="Image_Processing.html">Image processing</a>
    </li>
    <li>
      <a href="Manual_Annotation.html">Manual annotation</a>
    </li>
    <li>
      <a href="Normalization.html">Normalization</a>
    </li>
    <li>
      <a href="Spatial_Features.html">Spatial features</a>
    </li>
    <li>
      <a href="Neighborhood.html">Neighborhood Analysis</a>
    </li>
    <li>
      <a href="3D_visualization.html">3D visualization</a>
    </li>
    <li>
      <a href="Split_Data.html">Split Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="License.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ludvigla/STUtility_web_site">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Image processing</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-05-05
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>STUtility_web_site/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20191031code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20191031)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20191031code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20191031)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomludviglaSTUtilitywebsitetree8fe52086be051b10c604d5d7229c80f72abaefeatargetblank8fe5208a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/ludvigla/STUtility_web_site/tree/8fe52086be051b10c604d5d7229c80f72abaefea" target="_blank">8fe5208</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomludviglaSTUtilitywebsitetree8fe52086be051b10c604d5d7229c80f72abaefeatargetblank8fe5208a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/ludvigla/STUtility_web_site/tree/8fe52086be051b10c604d5d7229c80f72abaefea" target="_blank">8fe5208</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rproj.user/
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/manual_annotation.png
    Ignored:    pre_data/

Unstaged changes:
    Modified:   myproject.Rproj

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with <code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<style type="text/css">
div.main-container {
  background-color: #FFFFFF !important;
  max-width: 1800px;
  margin-left: auto;
  margin-right: auto;
}
</style>
<style>
#TOC {
  background: url("https://www.spatialresearch.org/wp-content/uploads/2019/09/str-logo-spatial_research_3@2x.png");
  background-size: contain;
  padding-top: 100px !important;
  background-repeat: no-repeat;
  width: 500px;
  color: black;
  border-color: #FFFFFF !important;
}
</style>
<style> code, pre{
  background-color: #F8F8F8 !important;
  color: black !important;
}
</style>
<style>
.watch-out { background-color: #F0F0F0; border: 1px #F0F0F0; }
<style>
<style> 
body {
  color: black;
  font-size: 20pt;
}
</style>
<pre class="r"><code>library(STutility)</code></pre>
<div id="loading-and-handling-images" class="section level1">
<h1>Loading and handling images</h1>
<hr />
<p>With the Seurat object created, we are ready to load the H&amp;E images from the file paths provided in the infoTable. The LoadImages() function allow you to load the images into the Seurat object and will automatically save a scaled down version of each image that you can use for plotting.</p>
<p>You should be aware that the resolution is by default very low (400 pixels wide). The first reason for this is that STutility makes it possible to apply various image transformations to the images, but since these processing steps are slow and requires significant amounts of memory, it is much faster to learn the transformations on a small image and later apply the transformations to a higher resolution image. The second reason is that drawing higher resoltuion images inside RStudio can be extremely slow and clumsy to work with interactively.</p>
<p>NOTE: If you ever need to reload your images or load them in higher resolution you must make sure that the image paths are correct. If you have moved the images or changed your working directory, reloading the images will not work.</p>
<p><br></p>
<pre class="r"><code>se &lt;- LoadImages(se, time.resolve = FALSE, verbose = TRUE)</code></pre>
<p><br> If you want to have a quick look at the sections you can draw them in the RStudio Viewer using the <code>ImagePlot</code> function. If you specify <code>method = &quot;raster&quot;</code> the images will be drawn as a plot instead. <br></p>
<pre class="r"><code>ImagePlot(se, method = &quot;raster&quot;, type = &quot;raw&quot;)</code></pre>
<p><img src="figure/Image_Processing.Rmd/plot_raw_he-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><br> OBS: We have noted that some systems experience memory problems due to large size image information not being cleared when removing old objects. If this is the case, the user can set <code>time.resolve = TRUE</code>, which will add a small garbage collection step to avoid this issue.</p>
<p>NOTE: time.resolve is set to TRUE as default in the current version. This will imply that the process takes approximately 5 sec longer per sample.</p>
<hr />
<p>The core of an ST experiment is the interplay between quantitative RNAseq data and qualitative image data. Working in R, the handling of image information and plotting of multiple samples can get a bit cumbersome, especially if the images are of higher resolution. Typically, we rarely find much use of plotting multiple samples together with the tissue images within this framework. Instead, the general tip is to perform the various plots using only the expression data, and after finding anything of interest that warrants further analysis - take out that sample and plot it individually together with the stained tissue image.</p>
<p>For example, we can visualize the number of RNA features and RNA counts on the ST array coordinates and quickly verify where we capture more material from the tissue.</p>
<p><br></p>
<pre class="r"><code># TODO: make sure that levels are set in facets

ST.FeaturePlot(object = se, 
               features = c(&quot;nFeature_RNA&quot;), 
               cols = c(&quot;lightgray&quot;, &quot;mistyrose&quot;, &quot;red&quot;, &quot;dark red&quot;, &quot;black&quot;),
               ncol = 2)</code></pre>
<p><img src="figure/Image_Processing.Rmd/unnamed-chunk-1-1.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ST.FeaturePlot(object = se, 
               features = c(&quot;nCount_RNA&quot;), 
               cols = c(&quot;lightgray&quot;, &quot;mistyrose&quot;, &quot;red&quot;, &quot;dark red&quot;, &quot;black&quot;),
               ncol = 2)</code></pre>
<p><img src="figure/Image_Processing.Rmd/unnamed-chunk-1-2.png" width="100%" style="display: block; margin: auto;" /> <br> To overlay a spatial feature heatmaps on top of the HE image we use the FeatureOverlay() function. Handling large images gets clumsy so we recommend to work with down-scaled versions of the images that takes up less memory. <br></p>
<pre class="r"><code>FeatureOverlay(se, 
               features = &quot;Cck&quot;, 
               pt.size = 1.5,
               cols = c(&quot;lightgray&quot;, &quot;mistyrose&quot;, &quot;red&quot;, &quot;dark red&quot;, &quot;black&quot;), 
               type = &quot;raw&quot;)</code></pre>
<p><img src="figure/Image_Processing.Rmd/overlay_features1-1.png" width="100%" style="display: block; margin: auto;" /> <br> You can see that by default, the first sample was selected for the feature overlay. If you wish to visualize another sample you can specify the <code>sample.index</code> parameter. If you wish to plot several features side by side, you can just specify as many as you want. The layout of the Overlay plots are controlled by the <code>ncols.samples</code> and the <code>ncols.features</code> parameters. <code>ncols.samples</code> can be used to adjust how many columns should be drawn for the sections in your Seurat object while the <code>ncols.features</code> controls how many columns should be drawn for the selected features. Here we are drawing 4 features on 1 section so if we want to draw the features a layout of 2x2 we can simply set <code>ncols.features = 2</code> <br></p>
<pre class="r"><code>FeatureOverlay(se, 
               features = c(&quot;Cck&quot;, &quot;Calb2&quot;, &quot;Dcn&quot;, &quot;Nrgn&quot;), 
               cols = c(&quot;lightgray&quot;, &quot;mistyrose&quot;, &quot;red&quot;, &quot;dark red&quot;, &quot;black&quot;), 
               type = &quot;raw&quot;,
               ncols = 2)</code></pre>
<p><img src="figure/Image_Processing.Rmd/overlay_features_2-1.png" width="100%" style="display: block; margin: auto;" /> <br> If you want to highlight only the regions with high feature values, you can add some opacity to the spots using the <code>add.alpha</code> parameter. Setting <code>add.alpha = TRUE</code> will add opacity to the spots scaled by the feature values, meaning that spots with high values will be visible and spots with low values will be transparent. <br></p>
<pre class="r"><code>FeatureOverlay(se, 
               features = c(&quot;Cck&quot;, &quot;Calb2&quot;, &quot;Dcn&quot;, &quot;Nrgn&quot;), 
               add.alpha = TRUE,
               cols = c(&quot;lightgray&quot;, &quot;mistyrose&quot;, &quot;red&quot;, &quot;dark red&quot;, &quot;black&quot;), 
               type = &quot;raw&quot;,
               ncols = 2)</code></pre>
<p><img src="figure/Image_Processing.Rmd/overlay_features_3-1.png" width="100%" style="display: block; margin: auto;" /> <br> Within the overlay area you also have a section number in the top left corner as well as a scale bar (showing the width of 500 microns) in the top right corner which can be disabled using the <code>sample.label</code> and <code>show.sb</code> options. Another thing we can do is to pass a custom theme (compatible with ggplot2 plots) if you wish to change the plot appearance.</p>
<p>Now we can for example disable the section number, scale bar, plot title and remove the legend from the plots. <br></p>
<pre class="r"><code>customtheme &lt;- theme(legend.position = &quot;None&quot;, plot.title = element_blank())

FeatureOverlay(se, 
               features = c(&quot;Cck&quot;, &quot;Calb2&quot;, &quot;Dcn&quot;, &quot;Nrgn&quot;), 
               cols = c(&quot;lightgray&quot;, &quot;mistyrose&quot;, &quot;red&quot;, &quot;dark red&quot;, &quot;black&quot;), 
               type = &quot;raw&quot;,
               ncols = 2,
               sample.label = FALSE,
               show.sb = FALSE) &amp; customtheme </code></pre>
<p><img src="figure/Image_Processing.Rmd/overlay_features_4-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><br> Another feature that can be useful for visualization purposes is to mask the background of the HE images. The function <code>MaskImages()</code> can be used to remove the background currently works well for tissues with a clearly defined border. Since masking is a non trivial problem, it is possible that the masking will fail for certain tissue types. If this happens you can try to modify the parameters in <code>MaskImages()</code> or create your own masking function to see if you can obtain better results (see <code>?MaskImages</code> for more details).</p>
<p>NOTE: the image procesing methods have been optimized for small images and if you have loaded images in higher resolution there is a greater risk that some steps will fail. If you still wish to use images with higher resolution there are instructions on how to do this in the Resolution section below. <br></p>
<pre class="r"><code>se &lt;- MaskImages(object = se)</code></pre>
<p><br> There are 5 types of images which can be stored into the Seurat object; “raw”, “masked”, “masked.masks”, “processed” and “processed.masks”. “raw” represents the original HE images without any modifications. In the “masked” images, the background have been removed using a binary mask which is stored under the name “masked.masks”. “processed” images have been further modified by applying rigid transformations such as rotations and shifts and these images also have the corresponding “processed.masks”.</p>
<p>You can look at the HE images using the ImagePlot() function.The ImagePlot function will automatically label the samples from 1 to N where N is the total number of samples. If you are running the function within RStudio you can take advantage of the Viewer panel to explore the images. Setting method=“raster” and the plot function will use the default plotting device in R. You can also change the layout of the images by forcing a different number of columns by setting ncols. <br></p>
<pre class="r"><code># Plot HE images
ImagePlot(se, ncols = 2, method = &quot;raster&quot;, type = &quot;raw&quot;) # Raw image</code></pre>
<p><img src="figure/Image_Processing.Rmd/image_types-1.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ImagePlot(se, ncols = 2, method = &quot;raster&quot;, type = &quot;masked&quot;) # Masked image</code></pre>
<p><img src="figure/Image_Processing.Rmd/image_types-2.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ImagePlot(se, ncols = 2, method = &quot;raster&quot;, type = &quot;masked.masks&quot;, darken = TRUE) # Mask</code></pre>
<p><img src="figure/Image_Processing.Rmd/image_types-3.png" width="100%" style="display: block; margin: auto;" /></p>
<div id="individual-transformations-to-images" class="section level3">
<h3>Individual transformations to images</h3>
<p>If you follow along the test set in this vignette, you will notice that the samples are not quite aligned with respect to each other on the array. We have implemented three different methods to align or process images; <code>AlignImages()</code>, <code>WarpImages()</code> and <code>ManualAlignImages()</code>.</p>
</div>
<div id="automatic-alignment" class="section level3">
<h3>Automatic alignment</h3>
<p>The automatic alignment method (<code>AlignImages()</code>) first tries to detect the tissue edges from each image. By default, the first image (<code>reference.index = 1</code>) is used as a reference but you can specify any sample with the <code>reference.index</code>option. Then, for each image, a transformation matrix is learned that can be used to map coordinates to the reference image. This alignment method will sometimes fail and in that case you can align the images manually instead using the <code>ManualAlignImages()</code> function. This is typically required when; (1) the tissue is bigger than the image and therefore outside the frame, (2) the tissue has a symmetrical shape (for example if the tissue has a round shape, it will be difficult to find an optimal alignment using the tissue edges), (3) the masking has failed.</p>
<p>The actual transformation is done using the <code>imwarp()</code> function from the imager R package using a “backward” transformation strategy. This method makes sure that every pixel is painted using linear interpolation so the aligned images will experience some quality loss.</p>
<p>The automatic alignment should only be used on tissue sections with the same morphology, so to demonstrate this we’ll look at the “2k” array mouse brain data. <br></p>
<pre class="r"><code># Look at the HE images from the &quot;2k&quot; array mouse brain dataset
ImagePlot(se, method = &quot;raster&quot;, type = &quot;raw&quot;, ncols = 4)</code></pre>
<p><img src="figure/Image_Processing.Rmd/2k_images-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>First we need to mask the images and then we can apply the automatic alignment method. As you can see, the automatic alignment method can apply rotations, translations and reflections to the images.</p>
<pre class="r"><code>se &lt;- se %&gt;% MaskImages() %&gt;% AlignImages()</code></pre>
<pre class="r"><code>ImagePlot(se, method = &quot;raster&quot;, type = &quot;processed&quot;, ncols = 4)</code></pre>
<p><img src="figure/Image_Processing.Rmd/align_images-1.png" width="1152" style="display: block; margin: auto;" /> <br></p>
<div id="manual-rigid-transformations" class="section level4">
<h4>Manual rigid transformations</h4>
<p>An alternative method is to apply rigid transformations to the images using the <code>WarpImages()</code> function. This function takes a list of transformations for the samples of your choice and converts the transformations into mapping functions which can be used to warp the images.</p>
<p>The available transformations are; “center.x”, “center.y”, “mirror.x”, “mirror.y” and “angle”</p>
<p>Remember that you need to mask the images before processing them! <br></p>
<pre class="r"><code>transforms &lt;- list(&quot;2&quot; = list(&quot;angle&quot; = 90))
se.rotate90 &lt;- WarpImages(se, transforms)
ImagePlot(se.rotate90, method = &quot;raster&quot;)</code></pre>
<p><img src="figure/Image_Processing.Rmd/warpimages-1.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>transforms &lt;- list(&quot;2&quot; = list(&quot;mirror.x&quot; = T))
se.mirrorx &lt;- WarpImages(se, transforms)
ImagePlot(se.mirrorx, method = &quot;raster&quot;)</code></pre>
<p><img src="figure/Image_Processing.Rmd/warpimages-2.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>transforms &lt;- list(&quot;2&quot; = list(&quot;mirror.y&quot; = T))
se.mirrory &lt;- WarpImages(se, transforms)
ImagePlot(se.mirrory, method = &quot;raster&quot;)</code></pre>
<p><img src="figure/Image_Processing.Rmd/warpimages-3.png" width="100%" style="display: block; margin: auto;" /> <br> Any transformations applied to the images will also be applied to the coordinates used to map gene expression values onto the images. If you want to draw the plots on other images you can pass one of “raw”, “masked”, “processed” to the <code>type</code> option in the <code>FeatureOverlay</code> function. <br></p>
<pre class="r"><code>heatmap.colors &lt;- c(&quot;lightgray&quot;, &quot;mistyrose&quot;, &quot;red&quot;, &quot;dark red&quot;, &quot;black&quot;)
ST.FeaturePlot(se, features = &quot;Cck&quot;, cols = heatmap.colors, ncol = 2)</code></pre>
<p><img src="figure/Image_Processing.Rmd/show_warp-1.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ST.FeaturePlot(se.rotate90, features = &quot;Cck&quot;, cols = heatmap.colors, ncol = 2)</code></pre>
<p><img src="figure/Image_Processing.Rmd/show_warp-2.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ST.FeaturePlot(se.mirrorx, features = &quot;Cck&quot;, cols = heatmap.colors, ncol = 2)</code></pre>
<p><img src="figure/Image_Processing.Rmd/show_warp-3.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ST.FeaturePlot(se.mirrory, features = &quot;Cck&quot;, cols = heatmap.colors, ncol = 2)</code></pre>
<p><img src="figure/Image_Processing.Rmd/show_warp-4.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="visualize-multiple-samples" class="section level3">
<h3>Visualize multiple samples</h3>
<p>So far we have only plotted section 1 using the <code>FeatureOverlay</code> function. If you want to draw feature heatmaps on multiple section at the same time you can do this by specifing several sections in the <code>sampleids</code> option <br></p>
<pre class="r"><code>FeatureOverlay(se, features = &quot;Nrgn&quot;, 
                sampleids = 1:2,
                cols = c(&quot;lightgray&quot;, &quot;mistyrose&quot;, &quot;red&quot;, &quot;dark red&quot;, &quot;black&quot;),
                ncols = 2)</code></pre>
<p><img src="figure/Image_Processing.Rmd/viz_multiple-1.png" width="100%" style="display: block; margin: auto;" /> <br> As mentioned before, we can change the <code>type</code> parameter in the <code>FeatureOverlay()</code> function to override the default use of processed images. <br></p>
<pre class="r"><code>FeatureOverlay(object = se, features = c(&quot;Nrgn&quot;, &quot;Cck&quot;), 
               cols = c(&quot;lightgray&quot;, &quot;mistyrose&quot;, &quot;red&quot;, &quot;dark red&quot;, &quot;black&quot;), 
               sampleids = 2,
               type = &quot;raw&quot;)</code></pre>
<p><img src="figure/Image_Processing.Rmd/overlay_multi_raw-1.png" width="100%" style="display: block; margin: auto;" /></p>
 
<hr />
<p style="text-align: center;">
A work by <a href="mailto: j.bergenstrahle@scilifelab.se">Joseph Bergenstråhle</a> and <a href="mailto: ludvig.larsson@scilifelab.se">Ludvig Larsson</a>
</p>
<!-- Add icon library -->
<p><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css"></p>
<!-- Add font awesome icons -->
<p style="text-align: center;">
<a href="https://www.spatialresearch.org" class="fa fa-beer"></a>
</p>
<p> </p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.5 (2021-03-31)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] STutility_0.1.0 ggplot2_3.3.3   Seurat_3.1.5    workflowr_1.6.2

loaded via a namespace (and not attached):
  [1] uuid_0.1-4              backports_1.1.6         systemfonts_0.2.1      
  [4] plyr_1.8.6              igraph_1.2.6            lazyeval_0.2.2         
  [7] sp_1.4-1                splines_4.0.5           crosstalk_1.1.1        
 [10] listenv_0.8.0           digest_0.6.27           foreach_1.5.0          
 [13] htmltools_0.5.1.1       viridis_0.5.1           magick_2.3             
 [16] tiff_0.1-5              gdata_2.18.0            fansi_0.4.2            
 [19] magrittr_2.0.1          cluster_2.1.1           doParallel_1.0.15      
 [22] ROCR_1.0-11             globals_0.14.0          gmodels_2.18.1         
 [25] jpeg_0.1-8.1            colorspace_2.0-0        blob_1.2.1             
 [28] ggrepel_0.9.1           xfun_0.13               dplyr_1.0.5            
 [31] crayon_1.4.1            jsonlite_1.7.2          zeallot_0.1.0          
 [34] survival_3.2-10         zoo_1.8-9               iterators_1.0.12       
 [37] ape_5.4-1               glue_1.4.2              gtable_0.3.0           
 [40] webshot_0.5.2           leiden_0.3.7            future.apply_1.7.0     
 [43] scales_1.1.1            DBI_1.1.0               miniUI_0.1.1.1         
 [46] Rcpp_1.0.6              viridisLite_0.3.0       xtable_1.8-4           
 [49] spData_0.3.5            units_0.6-6             reticulate_1.18        
 [52] spdep_1.1-3             rsvd_1.0.3              akima_0.6-2            
 [55] tsne_0.1-3              htmlwidgets_1.5.3       httr_1.4.2             
 [58] RColorBrewer_1.1-2      ellipsis_0.3.1          ica_1.0-2              
 [61] farver_2.1.0            pkgconfig_2.0.3         uwot_0.1.10            
 [64] deldir_0.1-25           utf8_1.2.1              labeling_0.4.2         
 [67] tidyselect_1.1.0        rlang_0.4.10            manipulateWidget_0.10.1
 [70] reshape2_1.4.4          later_1.1.0.1           munsell_0.5.0          
 [73] tools_4.0.5             dbscan_1.1-5            generics_0.1.0         
 [76] ggridges_0.5.3          evaluate_0.14           stringr_1.4.0          
 [79] fastmap_1.0.1           yaml_2.2.1              knitr_1.28             
 [82] fs_1.5.0                fitdistrplus_1.1-3      rgl_0.100.54           
 [85] purrr_0.3.4             RANN_2.6.1              readbitmap_0.1.5       
 [88] pbapply_1.4-3           future_1.21.0           nlme_3.1-152           
 [91] mime_0.10               ggiraph_0.7.7           compiler_4.0.5         
 [94] plotly_4.9.3            png_0.1-7               e1071_1.7-3            
 [97] Morpho_2.8              tibble_3.1.0            stringi_1.5.3          
[100] gdtools_0.2.2           lattice_0.20-41         Matrix_1.3-2           
[103] classInt_0.4-3          shinyjs_1.1             vctrs_0.3.7            
[106] pillar_1.5.1            LearnBayes_2.15.1       lifecycle_1.0.0        
[109] lmtest_0.9-38           RcppAnnoy_0.0.18        data.table_1.14.0      
[112] cowplot_1.1.1           irlba_2.3.3             Rvcg_0.19.1            
[115] raster_3.1-5            httpuv_1.5.2            patchwork_1.1.1        
[118] colorRamps_2.3          R6_2.5.0                imager_0.42.1          
[121] promises_1.2.0.1        KernSmooth_2.23-18      gridExtra_2.3          
[124] bmp_0.3                 parallelly_1.24.0       codetools_0.2-18       
[127] gtools_3.8.2            boot_1.3-27             MASS_7.3-53.1          
[130] assertthat_0.2.1        rprojroot_1.3-2         withr_2.4.1            
[133] sctransform_0.2.1       expm_0.999-4            parallel_4.0.5         
[136] grid_4.0.5              tidyr_1.1.3             coda_0.19-3            
[139] class_7.3-18            rmarkdown_2.1           Rtsne_0.15             
[142] git2r_0.27.1            sf_0.9-7                shiny_1.4.0.2          </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
