<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Quality control</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/darkly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">STUtility</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Getting started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="getting_started.html">Load data</a>
    </li>
    <li>
      <a href="Quality_Control.html">Quality Control</a>
    </li>
    <li>
      <a href="image_processing.html">Image processing</a>
    </li>
    <li>
      <a href="Manual_annotation.html">Manual annotation</a>
    </li>
    <li>
      <a href="normalization.html">Normalization</a>
    </li>
    <li>
      <a href="spatial_features.html">Spatial features</a>
    </li>
    <li>
      <a href="Neighborhood.html">Neighborhood Analysis</a>
    </li>
    <li>
      <a href="3D_visualization.html">3D visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ludvigla/STUtility_web_site">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Quality control</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-06-05
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>STUtility_web_site/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20191031code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20191031)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20191031code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20191031)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomludviglaSTUtilitywebsitetree42cb54d1b38dac80f6d200b171a63b74572c7fd2targetblank42cb54da"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/ludvigla/STUtility_web_site/tree/42cb54d1b38dac80f6d200b171a63b74572c7fd2" target="_blank">42cb54d</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomludviglaSTUtilitywebsitetree42cb54d1b38dac80f6d200b171a63b74572c7fd2targetblank42cb54da" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/ludvigla/STUtility_web_site/tree/42cb54d1b38dac80f6d200b171a63b74572c7fd2" target="_blank">42cb54d</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/manual_annotation.png
    Ignored:    analysis/visualization_3D.Rmd
    Ignored:    pre_data/

Unstaged changes:
    Modified:   analysis/_site.yml

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/Quality_Control.Rmd</code>) and HTML (<code>docs/Quality_Control.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/6606127a5e9d85d291d4cf9c954c98f3510f2288/docs/Quality_Control.html" target="_blank">6606127</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-05
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/f0ea9c1bc366c0bbd518093595d1a64b4de20d2f/docs/Quality_Control.html" target="_blank">f0ea9c1</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-05
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/651f315ce0f861caa5a5d7e818d80b5e8222475c/docs/Quality_Control.html" target="_blank">651f315</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-05
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/a6a0c5196bc8bba19aa69e682d70c93e6ea8a8a7/docs/Quality_Control.html" target="_blank">a6a0c51</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-05
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/f3c5cb4905a86c1ca0ee2075f732bbd1a8039058/docs/Quality_Control.html" target="_blank">f3c5cb4</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-05
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/f89df7e751dce8dbee625124aec4b2d4f5355db2/docs/Quality_Control.html" target="_blank">f89df7e</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-05
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/1e67a7c24c05f1bfbd7e009ab6e8e5683d524fcf/docs/Quality_Control.html" target="_blank">1e67a7c</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/f71b0c0e2040de522d8e14b55fa62514946cbe6b/docs/Quality_Control.html" target="_blank">f71b0c0</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ludvigla/STUtility_web_site/blob/90a45cc603252363a455842d2ed4425670c4b7d2/analysis/Quality_Control.Rmd" target="_blank">90a45cc</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
<td>
Fixed dataset in Load data
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/8a54a4dcff4330975d91f89675f62cd68d1db77b/docs/Quality_Control.html" target="_blank">8a54a4d</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/5e466eba685882478df06c6cb5302617b594d5f4/docs/Quality_Control.html" target="_blank">5e466eb</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ludvigla/STUtility_web_site/blob/4819d2b3e85a9101e87e4a9db21eee40d8b04715/analysis/Quality_Control.Rmd" target="_blank">4819d2b</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
<td>
update website
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/377408de6fd4608249fb09726c1567cb8e273320/docs/Quality_Control.html" target="_blank">377408d</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/ed54ffbef0e280f537fbc3fc181e0f7b4965cc2e/docs/Quality_Control.html" target="_blank">ed54ffb</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/f14518cca2b6f9b7036ca210bf7dd3960c062ec2/docs/Quality_Control.html" target="_blank">f14518c</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/efd885b87f34286687c1b0c4a951e4d4db228517/docs/Quality_Control.html" target="_blank">efd885b</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/ludvigla/STUtility_web_site/4f42429b56872cd53e2bf772a36a76fe5d2778d8/docs/Quality_Control.html" target="_blank">4f42429</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/ludvigla/STUtility_web_site/blob/3660612d2d42cb8373d862dd22152e65ca236547/analysis/Quality_Control.Rmd" target="_blank">3660612</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
<td>
update website
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<style type="text/css">
div.main-container {
  background-color: #000000 !important;
  max-width: 1400px;
  margin-left: auto;
  margin-right: auto;
}
</style>
<style>
#TOC {
  background: url("https://www.spatialresearch.org/wp-content/uploads/2019/09/str-logo-spatial_research_3@2x.png");
  background-size: contain;
  padding-top: 100px !important;
  background-repeat: no-repeat;
  op: 5%;
  opacity: 0.8;
  width: 500px;
  color: white;
  border-color: #000000 !important;
}
</style>
<style> code, pre{
  background-color: #000000 !important;
  color: white !important;
}
</style>
<style> 
body {
  color: white
}
</style>
<pre class="r"><code>library(STutility)
library(ggplot2)</code></pre>
<div id="quality-control" class="section level2">
<h2>Quality Control</h2>
<hr />
<p>Here we’ll go through some basic steps to assess the quality of your data and how to apply filters to remove low abundant genes and poor quality spots.</p>
<div id="include-all-spots" class="section level3">
<h3>Include all spots</h3>
<p>If you expect that you have over-permeabilized your tissue it could be useful to look at the expression patterns outside the tissue region as well. This can be done by loading the</p>
<p>Here we have a new infotable data.frame where the file paths in the “samples” column have been set to the &quot;*raw_feature_bc_matrix.h5&quot; matrrices instead of the filtered ones. Now we can load all spots into our Seurat object by setting <code>disable.subset = TRUE</code>.</p>
<pre class="r"><code>infoTable</code></pre>
<pre><code>                                                                                                     samples
1         ~/10x/Public_data/S1_Sagittal_posterior/V1_Mouse_Brain_Sagittal_Posterior_raw_feature_bc_matrix.h5
2         ~/10x/Public_data/S2_Sagittal_posterior/V1_Mouse_Brain_Sagittal_Posterior_raw_feature_bc_matrix.h5
3           ~/10x/Public_data/S1_Sagittal_anterior/V1_Mouse_Brain_Sagittal_Anterior_raw_feature_bc_matrix.h5
4 ~/10x/Public_data/S2_Sagittal_anterior/V1_Mouse_Brain_Sagittal_Anterior_Section_2_raw_feature_bc_matrix.h5
                                                            imgs
1 ~/10x/Public_data/S1_Sagittal_posterior/tissue_hires_image.png
2 ~/10x/Public_data/S2_Sagittal_posterior/tissue_hires_image.png
3  ~/10x/Public_data/S1_Sagittal_anterior/tissue_hires_image.png
4  ~/10x/Public_data/S2_Sagittal_anterior/tissue_hires_image.png
                                                          spotfiles
1 ~/10x/Public_data/S1_Sagittal_posterior/tissue_positions_list.csv
2 ~/10x/Public_data/S2_Sagittal_posterior/tissue_positions_list.csv
3  ~/10x/Public_data/S1_Sagittal_anterior/tissue_positions_list.csv
4  ~/10x/Public_data/S2_Sagittal_anterior/tissue_positions_list.csv
                                                            json
1 ~/10x/Public_data/S1_Sagittal_posterior/scalefactors_json.json
2 ~/10x/Public_data/S2_Sagittal_posterior/scalefactors_json.json
3  ~/10x/Public_data/S1_Sagittal_anterior/scalefactors_json.json
4  ~/10x/Public_data/S2_Sagittal_anterior/scalefactors_json.json</code></pre>
<pre class="r"><code>se &lt;- InputFromTable(infoTable, disable.subset = TRUE)</code></pre>
<p>The tissue borders are quite easy to see in the plot but you can also see that there have been transcripts captured also outside of the tissue. During library preparation, transcripts can diffuse out into the solution and end up anywhere outside the tissue but we can know from the TO experiments that the transcripts captured under the tissue form a cDNA footprint that accurately reflects the tissue morphology and that the transcripts have diffused vertically from the cells in the tissue down onto the capture area surface.</p>
<p>It can be good to keep this in mind when you see that you have holes in your tissue with no cells. You might detect quite a lot of transcripts in such holes and it is therefore important to carefully remove spots that are not covered by cells. If the automatic tissue detection algorithm run by spaceranger fails to find such holes, it could be a good idea to manually remove them using Loupe Browser before running spaceranger.</p>
<pre class="r"><code>ST.FeaturePlot(se, features = &quot;nFeature_RNA&quot;, dark.theme = TRUE, cols = c(&quot;dark blue&quot;, &quot;cyan&quot;, &quot;yellow&quot;, &quot;red&quot;, &quot;dark red&quot;))</code></pre>
<p><img src="figure/Quality_Control.Rmd/plot_all_spots-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-plot_all_spots-1">
Past versions of plot_all_spots-1.png
</button>
</p>
<div id="fig-plot_all_spots-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ludvigla/STUtility_web_site/blob/f71b0c0e2040de522d8e14b55fa62514946cbe6b/docs/figure/Quality_Control.Rmd/plot_all_spots-1.png" target="_blank">f71b0c0</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
</tr>
<tr>
<td>
<a href="https://github.com/ludvigla/STUtility_web_site/blob/4f42429b56872cd53e2bf772a36a76fe5d2778d8/docs/figure/Quality_Control.Rmd/plot_all_spots-1.png" target="_blank">4f42429</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Now let’s load the data with the subsetting enabled. Here we can use wither the raw matrices or the filtered matrices as long as we have spotfiles available in our infoTable data.frame which will be used to select the spots under tissue.</p>
<pre class="r"><code>se &lt;- InputFromTable(infoTable)</code></pre>
<p>Sometimes it can be a good idea to filter the data to remove low quality spots or low abundant genes. When running <code>InputFromTable</code>, spots with 0 counts will automatically be removed but you also have the option to filter the data directly using one of the following arguments:</p>
<ul>
<li>min.gene.count : sets a threshold for the minimum allowed UMI counts of a gene across the whole dataset
<ul>
<li>min.gene.spots : sets a threshold for the minimum allowed number of spots where a gene is detected cross the whole dataset</li>
<li>min.spot.feature.count : sets a threshold for the minimum allowed number of unique genes in a spot</li>
<li>min.spot.count : sets a threshold for the minimum allowed UMI counts in a spot</li>
<li>topN : subset the expression matrix to include only the topN most expressed genes</li>
</ul></li>
</ul>
<p>You can also apply filters when the Seurat obect has been created which gives you more freedom to explore what could be a good threshold. Below we have plotted some basic features that you can use to define your filtering thresholds when running <code>InputFromTable</code>.</p>
<pre class="r"><code>p1 &lt;- ggplot() +
  geom_histogram(data = se[[]], aes(nFeature_RNA), fill = &quot;red&quot;, alpha = 0.7, color = &quot;gray&quot;, bins = 50) +
  Seurat::DarkTheme() +
  ggtitle(&quot;Unique genes per spot&quot;)

p2 &lt;- ggplot() +
  geom_histogram(data = se[[]], aes(nCount_RNA), fill = &quot;red&quot;, alpha = 0.7, color = &quot;gray&quot;, bins = 50) +
  Seurat::DarkTheme() +
  ggtitle(&quot;Total counts per spots&quot;)

gene_attr &lt;- data.frame(nUMI = Matrix::rowSums(se@assays$RNA@counts), 
                        nSpots = Matrix::rowSums(se@assays$RNA@counts &gt; 0))
p3 &lt;- ggplot() +
  geom_histogram(data = gene_attr, aes(nUMI), fill = &quot;red&quot;, alpha = 0.7, color = &quot;gray&quot;, bins = 50) +
  Seurat::DarkTheme() +
  scale_x_log10() +
  ggtitle(&quot;Total counts per gene (log10 scale)&quot;)

p4 &lt;- ggplot() +
  geom_histogram(data = gene_attr, aes(nSpots), fill = &quot;red&quot;, alpha = 0.7, color = &quot;gray&quot;, bins = 50) +
  Seurat::DarkTheme() +
  ggtitle(&quot;Total spots per gene&quot;)

cowplot::plot_grid(p1, p2, p3, p4)</code></pre>
<p><img src="figure/Quality_Control.Rmd/qc_plots-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-qc_plots-1">
Past versions of qc_plots-1.png
</button>
</p>
<div id="fig-qc_plots-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ludvigla/STUtility_web_site/blob/f71b0c0e2040de522d8e14b55fa62514946cbe6b/docs/figure/Quality_Control.Rmd/qc_plots-1.png" target="_blank">f71b0c0</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
</tr>
<tr>
<td>
<a href="https://github.com/ludvigla/STUtility_web_site/blob/4f42429b56872cd53e2bf772a36a76fe5d2778d8/docs/figure/Quality_Control.Rmd/qc_plots-1.png" target="_blank">4f42429</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="filter-out-spots" class="section level3">
<h3>Filter out spots</h3>
<p>Let’s say that we want to remove all spots with fewer than 500 unique genes we can simply subset the using the <code>SubsetSTData</code> function and an expression.</p>
<p>NOTE: The Seurat package provides a <code>subset</code> method for Seurat objects but unfotunately this method will not work when using STUtility.</p>
<pre class="r"><code>se.subset &lt;- SubsetSTData(se, expression = nFeature_RNA &gt; 500)

cat(&quot;Spots removed: &quot;, ncol(se) - ncol(se.subset), &quot;\n&quot;)</code></pre>
<pre><code>Spots removed:  42 </code></pre>
</div>
<div id="mitochondrial-content" class="section level3">
<h3>Mitochondrial content</h3>
<p>It can also be useful to explore other features of the dataset to use for filtering, for example mitochondrial transcript content or ribosomal protein coding transcript content. Mitochondrial genes are prefixed with “mt-” in MGI nomenclature so we can collect these genes and then calculate the percentage of mitochondrial content per spot and add this information to our meta.data.</p>
<pre class="r"><code># Collect all genes coded on the mitochondrial genome
mt.genes &lt;- grep(pattern = &quot;^mt-&quot;, x = rownames(se), value = TRUE)
se$percent.mito &lt;- (Matrix::colSums(se@assays$RNA@counts[mt.genes, ])/Matrix::colSums(se@assays$RNA@counts))*100

# Collect all genes coding for ribosomal proteins
rp.genes &lt;- grep(pattern = &quot;^Rpl|^Rps&quot;, x = rownames(se), value = TRUE)
se$percent.ribo &lt;- (Matrix::colSums(se@assays$RNA@counts[rp.genes, ])/Matrix::colSums(se@assays$RNA@counts))*100

ST.FeaturePlot(se, features = &quot;percent.mito&quot;, dark.theme = TRUE, cols = c(&quot;dark blue&quot;, &quot;cyan&quot;, &quot;yellow&quot;, &quot;red&quot;, &quot;dark red&quot;))</code></pre>
<p><img src="figure/Quality_Control.Rmd/subset_by_mt-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-subset_by_mt-1">
Past versions of subset_by_mt-1.png
</button>
</p>
<div id="fig-subset_by_mt-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ludvigla/STUtility_web_site/blob/f71b0c0e2040de522d8e14b55fa62514946cbe6b/docs/figure/Quality_Control.Rmd/subset_by_mt-1.png" target="_blank">f71b0c0</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
</tr>
<tr>
<td>
<a href="https://github.com/ludvigla/STUtility_web_site/blob/4f42429b56872cd53e2bf772a36a76fe5d2778d8/docs/figure/Quality_Control.Rmd/subset_by_mt-1.png" target="_blank">4f42429</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>ST.FeaturePlot(se, features = &quot;percent.ribo&quot;, dark.theme = TRUE, cols = c(&quot;dark blue&quot;, &quot;cyan&quot;, &quot;yellow&quot;, &quot;red&quot;, &quot;dark red&quot;))</code></pre>
<p><img src="figure/Quality_Control.Rmd/subset_by_mt-2.png" width="100%" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-subset_by_mt-2">
Past versions of subset_by_mt-2.png
</button>
</p>
<div id="fig-subset_by_mt-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/ludvigla/STUtility_web_site/blob/f71b0c0e2040de522d8e14b55fa62514946cbe6b/docs/figure/Quality_Control.Rmd/subset_by_mt-2.png" target="_blank">f71b0c0</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
</tr>
<tr>
<td>
<a href="https://github.com/ludvigla/STUtility_web_site/blob/4f42429b56872cd53e2bf772a36a76fe5d2778d8/docs/figure/Quality_Control.Rmd/subset_by_mt-2.png" target="_blank">4f42429</a>
</td>
<td>
Ludvig Larsson
</td>
<td>
2020-06-04
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We can also combine different thresholds to filter the data. Let’s say that we want to remove all spots with fewer than 500 unique genes and also spots with a high mitochondrial transcript content (&gt;30%).</p>
<pre class="r"><code># Keep spots with more than 500 unique genes and less than 30% mitochondrial content
se.subset &lt;- SubsetSTData(se, expression = nFeature_RNA &gt; 500 &amp; percent.mito &lt; 30)

cat(&quot;Spots removed: &quot;, ncol(se) - ncol(se.subset), &quot;\n&quot;)</code></pre>
<pre><code>Spots removed:  180 </code></pre>
</div>
<div id="removing-genes" class="section level3">
<h3>Removing genes</h3>
<p>If you have good reson to remove a certain type of gene, this can also be done quite easily as well. For example, you might want to keep only protein coding genes in your dataset. Here we demonstrate how to subset a Seurat object to include only protein coding genes using our predefined covnersion table, but you could also get this information elsewhere, e.g. bioMart.</p>
<pre class="r"><code>ensids &lt;- read.table(file = list.files(system.file(&quot;extdata&quot;, package = &quot;STutility&quot;), full.names = T, pattern = &quot;mouse_genes&quot;), header = T, sep = &quot;\t&quot;, stringsAsFactors = F)

# Print available biotypes
unique(ensids$gene_type)
keep.genes &lt;- subset(ensids, gene_type %in% &quot;protein_coding&quot;)$gene_name

# Subset Seurat object
se.subset &lt;- se[intersect(rownames(se), keep.genes), ]

cat(&quot;Number of genes removed : &quot;, nrow(se) - nrow(se.subset), &quot;\n&quot;)</code></pre>
 
<hr />
<p style="text-align: center;">
A work by <a href="j.bergenstrahle@scilifelab.se">Joseph Bergenstråhle</a> and <a href="ludvig.larsson@scilifelab.se">Ludvig Larsson</a>
</p>
<!-- Add icon library -->
<p><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css"></p>
<!-- Add font awesome icons -->
<p style="text-align: center;">
<a href="https://www.spatialresearch.org" class="fa fa-beer"></a>
</p>
<p> </p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.0 (2020-04-24)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

Random number generation:
 RNG:     Mersenne-Twister 
 Normal:  Inversion 
 Sample:  Rounding 
 
locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] STutility_0.1.0             ggplot2_3.3.0              
 [3] SingleCellExperiment_1.10.1 SummarizedExperiment_1.18.1
 [5] DelayedArray_0.14.0         matrixStats_0.56.0         
 [7] Biobase_2.48.0              GenomicRanges_1.40.0       
 [9] GenomeInfoDb_1.24.0         IRanges_2.22.1             
[11] S4Vectors_0.26.0            BiocGenerics_0.34.0        
[13] Seurat_3.1.5                workflowr_1.6.2            

loaded via a namespace (and not attached):
  [1] reticulate_1.15         tidyselect_1.0.0        htmlwidgets_1.5.1      
  [4] grid_4.0.0              Rtsne_0.15              munsell_0.5.0          
  [7] codetools_0.2-16        ica_1.0-2               units_0.6-6            
 [10] future_1.17.0           miniUI_0.1.1.1          withr_2.2.0            
 [13] colorspace_1.4-1        knitr_1.28              uuid_0.1-4             
 [16] ROCR_1.0-11             tensor_1.5              listenv_0.8.0          
 [19] labeling_0.3            git2r_0.27.1            GenomeInfoDbData_1.2.3 
 [22] polyclip_1.10-0         farver_2.0.3            rprojroot_1.3-2        
 [25] coda_0.19-3             LearnBayes_2.15.1       vctrs_0.3.0            
 [28] xfun_0.13               R6_2.4.1                doParallel_1.0.15      
 [31] rsvd_1.0.3              Morpho_2.8              ggiraph_0.7.0          
 [34] manipulateWidget_0.10.1 bitops_1.0-6            spatstat.utils_1.17-0  
 [37] assertthat_0.2.1        promises_1.1.0          scales_1.1.0           
 [40] imager_0.42.1           gtable_0.3.0            npsurv_0.4-0.1         
 [43] globals_0.12.5          bmp_0.3                 goftest_1.2-2          
 [46] rlang_0.4.6             zeallot_0.1.0           akima_0.6-2            
 [49] systemfonts_0.2.1       splines_4.0.0           lazyeval_0.2.2         
 [52] rgl_0.100.54            yaml_2.2.1              reshape2_1.4.4         
 [55] abind_1.4-5             crosstalk_1.1.0.1       backports_1.1.6        
 [58] httpuv_1.5.2            tools_4.0.0             spData_0.3.5           
 [61] ellipsis_0.3.0          raster_3.1-5            RColorBrewer_1.1-2     
 [64] Rvcg_0.19.1             ggridges_0.5.2          Rcpp_1.0.4.6           
 [67] plyr_1.8.6              zlibbioc_1.34.0         classInt_0.4-3         
 [70] purrr_0.3.4             RCurl_1.98-1.2          rpart_4.1-15           
 [73] dbscan_1.1-5            deldir_0.1-25           viridis_0.5.1          
 [76] pbapply_1.4-2           cowplot_1.0.0           zoo_1.8-8              
 [79] ggrepel_0.8.2           cluster_2.1.0           colorRamps_2.3         
 [82] fs_1.4.1                magrittr_1.5            data.table_1.12.8      
 [85] magick_2.3              readbitmap_0.1.5        gmodels_2.18.1         
 [88] lmtest_0.9-37           RANN_2.6.1              whisker_0.4            
 [91] fitdistrplus_1.0-14     patchwork_1.0.0         shinyjs_1.1            
 [94] lsei_1.2-0.1            mime_0.9                evaluate_0.14          
 [97] xtable_1.8-4            jpeg_0.1-8.1            gridExtra_2.3          
[100] compiler_4.0.0          tibble_3.0.1            KernSmooth_2.23-17     
[103] crayon_1.3.4            htmltools_0.4.0         mgcv_1.8-31            
[106] later_1.0.0             spdep_1.1-3             tiff_0.1-5             
[109] tidyr_1.0.3             expm_0.999-4            DBI_1.1.0              
[112] MASS_7.3-51.6           sf_0.9-3                boot_1.3-25            
[115] Matrix_1.2-18           gdata_2.18.0            igraph_1.2.5           
[118] pkgconfig_2.0.3         sp_1.4-1                plotly_4.9.2.1         
[121] xml2_1.3.2              foreach_1.5.0           webshot_0.5.2          
[124] XVector_0.28.0          stringr_1.4.0           digest_0.6.25          
[127] sctransform_0.2.1       RcppAnnoy_0.0.16        tsne_0.1-3             
[130] spatstat.data_1.4-3     rmarkdown_2.1           leiden_0.3.3           
[133] uwot_0.1.8              gdtools_0.2.2           gtools_3.8.2           
[136] shiny_1.4.0.2           lifecycle_0.2.0         nlme_3.1-147           
[139] jsonlite_1.6.1          viridisLite_0.3.0       pillar_1.4.4           
[142] lattice_0.20-41         fastmap_1.0.1           httr_1.4.1             
[145] survival_3.1-12         glue_1.4.0              spatstat_1.63-3        
[148] png_0.1-7               iterators_1.0.12        class_7.3-17           
[151] stringi_1.4.6           dplyr_0.8.5             irlba_2.3.3            
[154] e1071_1.7-3             future.apply_1.5.0      ape_5.3                </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
