<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="" />


<title>Spatial Features</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/darkly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">STUtility</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="installation.html">Installation</a>
</li>
<li>
  <a href="getting_started.html">Getting Started</a>
</li>
<li>
  <a href="image_processing.html">Image processing</a>
</li>
<li>
  <a href="normalization.html">Normalization</a>
</li>
<li>
  <a href="spatial_features.html">Spatial features</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ludvigla/STUtility_web_site">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Spatial Features</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-01-12
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>STUtility_web_site/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.5.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20191031code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20191031)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20191031code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20191031)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomludviglaSTUtilitywebsitetreed936ce7287f396c301ecbc32d982cce6df49f7a5targetblankd936ce7a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/ludvigla/STUtility_web_site/tree/d936ce7287f396c301ecbc32d982cce6df49f7a5" target="_blank">d936ce7</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomludviglaSTUtilitywebsitetreed936ce7287f396c301ecbc32d982cce6df49f7a5targetblankd936ce7a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/getting_started_cache/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with <code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<style type="text/css">
div.main-container {
  background-color: #000000 !important;
  max-width: 1400px;
  margin-left: auto;
  margin-right: auto;
}
</style>
<style>
#TOC {
  background: url("https://www.spatialresearch.org/wp-content/uploads/2019/09/str-logo-spatial_research_3@2x.png");
  background-size: contain;
  padding-top: 100px !important;
  background-repeat: no-repeat;
  op: 5%;
  opacity: 0.8;
  width: 500px;
  color: white;
  border-color: #000000 !important;
}
</style>
<style> code, pre{
  background-color: #000000 !important;
  color: white !important;
}
</style>
<style> 
body {
  color: white
}
</style>
<style>
.list-group-item.active, .list-group-item.active:focus, .list-group-item.active:hover {
    background-color: #375a7f;
}
</style>
<div id="finding-spatial-expression-patterns" class="section level2">
<h2>Finding spatial expression patterns</h2>
<p>The strength of untargeted whole transcriptome capture is the ability to perform unsupervised analysis and the ability to find spatial gene expression patterns. We’ve found good use of using non-negative matrix factorization (NMF) to find underlying patterns of transcriptomic profiles. This factor analysis, along with various dimensionality reduction techniques, can all be ran via “RunXXX()”, where X = the method of choice, e.g.: <br></p>
<pre class="r"><code>se &lt;- RunNMF(se, nfactors = 40) #Specificy nfactors to choose the number of factors, default=20.</code></pre>
<p><br> While <code>RunNMF()</code> is an STUtility add-on, others are supported via Seurat (<code>RunPCA()</code>, <code>RunTSNE</code>, <code>RunICA()</code>, <code>runUMAP()</code> ) and for all of them, the output are stored in the Seurat object.</p>
<p>We can then plot a variable number of dimensions across the samples. <br></p>
<pre class="r"><code>cscale &lt;- c(&quot;black&quot;, &quot;dark blue&quot;, &quot;cyan&quot;, &quot;yellow&quot;, &quot;red&quot;, &quot;dark red&quot;)

ST.DimPlot(se, 
           dims = 1:10,
           ncol = 2, # Sets the number of columns at dimensions level
           grid.ncol = 2, # Sets the number of columns at sample level
           reduction = &quot;NMF&quot;, 
           dark.theme = T, 
           pt.size = 0.5, 
           center.zero = F, 
           cols = cscale)</code></pre>
<p><img src="figure/spatial_features.Rmd/dimplot-1.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ST.DimPlot(se, 
           dims = 11:20,
           ncol = 2, 
           grid.ncol = 2, 
           reduction = &quot;NMF&quot;, 
           dark.theme = T, 
           pt.size = 0.5, 
           center.zero = F, 
           cols = cscale)</code></pre>
<p><img src="figure/spatial_features.Rmd/dimplot-2.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ST.DimPlot(se, 
           dims = 21:30,
           ncol = 2, 
           grid.ncol = 2, 
           reduction = &quot;NMF&quot;, 
           dark.theme = T, 
           pt.size = 0.5, 
               center.zero = F, 
           cols = cscale)</code></pre>
<p><img src="figure/spatial_features.Rmd/dimplot-3.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ST.DimPlot(se, 
           dims = 31:40,
           ncol = 2, 
           grid.ncol = 2, 
           reduction = &quot;NMF&quot;, 
           dark.theme = T, 
           pt.size = 0.5, 
               center.zero = F, 
           cols = cscale)</code></pre>
<p><img src="figure/spatial_features.Rmd/dimplot-4.png" width="100%" style="display: block; margin: auto;" /> <br> To extract the genes that drives the separation according to the dimensionality reduction, we can use the <code>ProjectDim()</code> function. <br></p>
<pre class="r"><code>ProjectDim(se, reduction = &quot;NMF&quot;, dims = c(2,4))</code></pre>
<pre><code>factor_ 2 
Positive:  Gng4, Synpr, Pcp4l1, Gad1, Gpsm1, Pcbp3, Meis2, Ptpro, Shisa8, Dlx1 
       Cpne4, Pbx3, Inpp5j, Cpne6, Tshz1, Sp8, Doc2g, Dcx, Lgr6, Rasl11b 
Negative:  Plp1, Pvalb, Mbp, Ptgds, Ttr, Pcp2, Trf, Calb1, Nrgn, Bc1 
       Sst, Mobp, Gng13, Itpr1, Cck, Ppp1r1b, Vsnl1, Fth1, Hpca, Enc1 
factor_ 4 
Positive:  Nptxr, Lypd1, Nov, Olfm1, Trh, Slc30a3, Ly6h, Hap1, Crym, Enc1 
       Hpcal1, Ak5, Cck, Atp2b4, Bhlhe22, Gda, Rtn1, Nnat, Pcdh8, Camk2d 
Negative:  Mbp, Plp1, Mobp, Trf, Pvalb, Apod, Hba-a1, Hbb-bs, Hba-a2, Pcp2 
       Cnp, Mag, Mal, Pcp4, Plekhb1, Hbb-bt, Cryab, Sept4, Qdpr, Cldn11 </code></pre>
<pre><code>An object of class Seurat 
37188 features across 12125 samples within 2 assays 
Active assay: SCT (18594 features)
 1 other assay present: RNA
 2 dimensional reductions calculated: NMF, umap</code></pre>
</div>
<div id="clustering" class="section level2">
<h2>Clustering</h2>
<p>Clustering is a standard procedure in genomic analysis, and the methods for doing so are numerous. Here we demonstrate an example where we use the result of the factor analysis the previous section. Going through the list of factors (e.g. via <code>ST:DimPlot(se, dims = [dims you want to look at])</code>), we can notice dimensions that are “spatially active”, i.e. that seems to confer a spatial pattern along their axis. We can extract these dimensions and use as input to e.g. clustering functions. Here, we use all dimensions from the NMF and construct a Shared Nearest Neighbor (SSN) Graph. <br></p>
<pre class="r"><code>se &lt;- FindNeighbors(object = se, verbose = FALSE, reduction = &quot;NMF&quot;, dims = 1:40)</code></pre>
<p><br> Followed by clustering using a modularity optimizer <br></p>
<pre class="r"><code>se &lt;- FindClusters(object = se, verbose = FALSE)</code></pre>
<p><br> And plotting of the clusters spatially <br></p>
<pre class="r"><code>library(RColorBrewer)
n &lt;- 19
qual_col_pals = brewer.pal.info[brewer.pal.info$category == &#39;qual&#39;,]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))

ST.FeaturePlot(object = se, features = &quot;seurat_clusters&quot;, dark.theme = T, cols = col_vector, pt.size = 0.5)</code></pre>
<p><img src="figure/spatial_features.Rmd/plot_clusters-1.png" width="100%" style="display: block; margin: auto;" /> <br> If you think that the distribution of clusters gets too cluttered, you can also split the view so that only one cluster at the time gets colored. <br></p>
<pre class="r"><code>ST.FeaturePlot(object = se, features = &quot;seurat_clusters&quot;, dark.theme = T, pt.size = 1, split.labels = T, indices = 1)</code></pre>
<p><img src="figure/spatial_features.Rmd/plot_clusters_split-1.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ST.FeaturePlot(object = se, features = &quot;seurat_clusters&quot;, dark.theme = T, pt.size = 1, split.labels = T, indices = 3)</code></pre>
<p><img src="figure/spatial_features.Rmd/plot_clusters_split-2.png" width="100%" style="display: block; margin: auto;" /> <br></p>
</div>
<div id="most-variable-features" class="section level2">
<h2>Most variable features</h2>
<p>We can take a specific look at some of the most variable features. <br></p>
<pre class="r"><code>head(se@assays$SCT@var.features, 20)</code></pre>
<pre><code> [1] &quot;Ttr&quot;           &quot;Pcp2&quot;          &quot;Ptgds&quot;         &quot;Hbb-bs&quot;       
 [5] &quot;Hba-a1&quot;        &quot;Plp1&quot;          &quot;Car8&quot;          &quot;Hba-a2&quot;       
 [9] &quot;Mbp&quot;           &quot;S100a5&quot;        &quot;Hbb-bt&quot;        &quot;Penk&quot;         
[13] &quot;Pvalb&quot;         &quot;Enpp2&quot;         &quot;Sst&quot;           &quot;Ppp1r1b&quot;      
[17] &quot;Gpr88&quot;         &quot;1500015O10Rik&quot; &quot;Mobp&quot;          &quot;Pcp4&quot;         </code></pre>
<pre class="r"><code>top &lt;- se@assays$SCT@var.features

fts &lt;- c(&quot;Ttr&quot;, &quot;Mbp&quot;, &quot;Nrgn&quot;)
for (ftr in fts) {
  FeatureOverlay(se, 
                    features = ftr, 
                    sampleids = 1:2,
                    cols = c(&quot;black&quot;, &quot;dark blue&quot;, &quot;cyan&quot;, &quot;yellow&quot;, &quot;red&quot;, &quot;dark red&quot;),
                    pt.size = 0.5, 
                    pt.alpha = 0.5, 
                    dark.theme = T)
}</code></pre>
<p><br></p>
</div>
<div id="compare-graph-embeddings" class="section level2">
<h2>Compare graph embeddings</h2>
<p>Another useful feature is that you can now compare the spatial distribution of a gene with the typical “graph embeddings” s.a. UMAP and t-SNE. <br></p>
<pre class="r"><code># Run UMAP
se &lt;- RunUMAP(se, reduction = &quot;NMF&quot;, n.neighbors = 10)</code></pre>
<pre class="r"><code># Define colors for heatmap
heatmap.colors &lt;- c(rgb(40, 40, 40, maxColorValue = 255), &quot;dark blue&quot;, &quot;cyan&quot;, &quot;white&quot;)
fts &lt;- c(&quot;Prkcd&quot;, &quot;Opalin&quot;, &quot;Lamp5&quot;)

# plot transformed features expression on UMAP embedding
p.fts &lt;- lapply(fts, function(ftr) {
  FeaturePlot(se, features = ftr, reduction = &quot;umap&quot;, order = TRUE, cols = heatmap.colors) + DarkTheme()
})

# plot transformed features expression on Visium coordinates
p3 &lt;- ST.FeaturePlot(se, features = fts, ncol = 2, grid.ncol = 1, cols = heatmap.colors, pt.size = 0.5, dark.theme = T)

# Construct final plot
cowplot::plot_grid(cowplot::plot_grid(plotlist = p.fts, ncol = 1), p3, ncol = 2, rel_widths = c(1, 1))</code></pre>
<p><img src="figure/spatial_features.Rmd/embedding_vs_ST-1.png" width="864" style="display: block; margin: auto;" /></p>
</div>
<div id="rgb-dimensionality-reduction-plots" class="section level2">
<h2>RGB dimensionality reduction plots</h2>
<p>One approach to visualize the result of dimensionality reduction is to use the first three dimensions and transform the values into RGB color space. This 3 dimensional space can then be utilized for spatial visualization. We demonstrate the technique with UMAP, using our factors as input: <br></p>
<pre class="r"><code>se &lt;- RunUMAP(object = se, dims = 1:40, verbose = FALSE, n.components = 3, reduction = &quot;NMF&quot;, reduction.name = &quot;umap.3d&quot;)</code></pre>
<p><br> We use the first three dimensions for plotting: <br></p>
<pre class="r"><code>ST.DimPlot(object = se, dims = 1:3, reduction = &quot;umap.3d&quot;, blend = T, dark.theme = T, pt.size = 0.5)</code></pre>
<p><img src="figure/spatial_features.Rmd/UMAP_blend-1.png" width="960" style="display: block; margin: auto;" /></p>
<p><br> Once again the <code>ProjectDim()</code> function can be used to display the genes that are most strongly correlated with the coordinate system. Note in the function call above that we defined <code>reduction.name</code>, which are subsequently stored in the Seurat object in the reduction slot: <br></p>
<pre class="r"><code>ProjectDim(se, reduction = &quot;umap.3d&quot;)</code></pre>
<pre><code>umap.3d_ 1 
Positive:  Mbp, Plp1, Pvalb, Pcp2, Mobp, Trf, Cbln1, Cbln3, Car8, Gabra6 
       Apod, Zic1, Cnp, Mag, Sparc, Gm2694, Plekhb1, S100b, Ptgds, Sept4 
Negative:  Nrgn, Camk2n1, Ctxn1, Cck, Camk2a, Olfm1, Basp1, Nptxr, Chn1, Sst 
       Arpp21, Ddn, Enc1, Hpcal4, Gpr88, Penk, Calm2, Rtn1, Mef2c, Snca 
umap.3d_ 2 
Positive:  Calb2, Pcp2, Car8, Fabp7, Pvalb, Doc2g, Gng4, Gng13, Cbln1, Ppp1r17 
       Cdhr1, Zic1, Slc1a3, Homer3, Calb1, Cbln3, Dner, Rgs8, Gabra6, S100a5 
Negative:  Ppp1r1b, Gpr88, Penk, Mbp, Plp1, Pde10a, Tac1, Rgs9, Pde1b, Mobp 
       Hpca, Scn4b, Adcy5, Nrgn, Rasd2, Adora2a, Arpp21, Mal, Ngef, Gng7 
umap.3d_ 3 
Positive:  Nrgn, Cck, Slc17a7, Pcp2, Olfm1, 3110035E14Rik, Cbln1, Cbln3, Camk2n1, Atp1a1 
       Adcy1, Snap25, Car8, Gabra6, 1110008P14Rik, Stx1a, Gng13, Itpr1, Gm2694, Pvalb 
Negative:  Penk, Ppp1r1b, Gpr88, Ttr, Pde10a, Plp1, Mbp, Tac1, Pde1b, Rgs9 
       Pcp4, Scn4b, Adcy5, Mobp, Rasd2, Adora2a, Pcp4l1, Nnat, Drd2, Gng7 </code></pre>
<pre><code>An object of class Seurat 
37188 features across 12125 samples within 2 assays 
Active assay: SCT (18594 features)
 1 other assay present: RNA
 3 dimensional reductions calculated: NMF, umap, umap.3d</code></pre>
<p><br></p>
</div>
<div id="dea" class="section level2">
<h2>DEA</h2>
<p>Lets try this out by an example. Looking at , lets say we are interested in cluster 19, and want to answer - “Which genes are significantly up-regulated in this region compared to the overall expression across the tissue?”</p>
<pre class="r"><code>markers &lt;- FindMarkers(se, ident.1 = &quot;19&quot;)
head(markers)</code></pre>
<pre><code>     p_val avg_logFC pct.1 pct.2 p_val_adj
Dlx1     0  1.525068 0.993 0.147         0
Dcx      0  1.410173 0.975 0.141         0
Sp9      0  1.179423 0.942 0.132         0
Sp8      0  1.142941 0.892 0.061         0
Lgr6     0  1.065057 0.813 0.071         0
Dlx2     0  1.054816 0.896 0.098         0</code></pre>
<p>Note that the clusters were already set as the Seurat objects levels. Type <code>levels(se)</code> to see the current levels of your object. If other clusters, annotations etc are of interest, set this before by specifying <code>Idents(se) &lt;-</code></p>
<p>Note also, if we are interested in comparing two levels against each other, and not just “one against the rest”, we simply add a <code>ident.2 =</code> parameter to the above.</p>
<pre class="r"><code>FeatureOverlay(se, features = &quot;Dlx1&quot;, 
                    sampleids = 1:2,
                    cols = c(&quot;black&quot;, &quot;dark blue&quot;, &quot;cyan&quot;, &quot;yellow&quot;, &quot;red&quot;, &quot;dark red&quot;),
                    pt.size = 1, 
                    pt.alpha = 0.5,
                    dark.theme = T)</code></pre>
<p><img src="figure/spatial_features.Rmd/unnamed-chunk-2-1.png" width="576" style="display: block; margin: auto;" /> <br></p>
</div>
<div id="spatial-auto-correlation" class="section level2">
<h2>Spatial Auto-correlation</h2>
<p>STUtility also includes an additional method for finding genes with spatial patterns across the tissue. The ranking method makes use neighborhood networks to compute the spatial lag for each gene, here defined as the summed expression of that gene in neighboring spots. Each gene is then ranked by the correlation between the lag vector and the original expression vector. The output is a ranked list of the genes with highest correlation between the two vectors.</p>
<pre class="r"><code>SpatialGenes(se)</code></pre>
</div>
<div id="region-neighbors" class="section level2">
<h2>Region neighbors</h2>
<p>To demonstrate the regional neighbors feature, we will use the Breast Cancer data available on the 10x Genomics <a href="https://support.10xgenomics.com/spatial-gene-expression/datasets">website</a>.</p>
<div id="connected-spatial-network" class="section level3">
<h3>Connected Spatial Network</h3>
<p>If you have defined a region of interest and you want to find all spots neighboring to this region you can use the RegionNeighbours() function to automatically detect such spots. For example, let’s say that we want to select all neighbors to cluster 12 in the plot below and find the differential expressed markers between the neighbors.</p>
<pre class="r"><code>cols &lt;- c(&quot;#771155&quot;, &quot;#AA4488&quot;, &quot;#CC99BB&quot;, &quot;#114477&quot;, &quot;#4477AA&quot;, &quot;#77AADD&quot;, &quot;#117777&quot;, &quot;#44AAAA&quot;, &quot;#77CCCC&quot;, &quot;#117744&quot;, &quot;#44AA77&quot;, &quot;#88CCAA&quot;, &quot;#777711&quot;, &quot;#AAAA44&quot;, &quot;#DDDD77&quot;, &quot;#774411&quot;, &quot;#AA7744&quot;, &quot;#DDAA77&quot;)

ST.FeaturePlot(se, features = &quot;seurat_clusters&quot;, pt.size = 1, cols=cols, dark.theme = T)</code></pre>
<p><img src="figure/spatial_features.Rmd/unnamed-chunk-4-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>First, RegionNeighbours() function to find the neighboring spots but we have to make sure that the active identity is appropriately set, here we use ‘seurat_clusters’.</p>
<pre class="r"><code>se &lt;- SetIdent(se, value = &quot;seurat_clusters&quot;)
se &lt;- RegionNeighbours(se, id = 2, verbose = TRUE)</code></pre>
<pre><code>Creating Connected Netork using KNN ... 
Found 4763 neighbours for id 2 ... 
Excluding neighbours from the same group ... 
569 neighbours left ... 
Naming all neighbours nbs_2 ... 
Saving neighbour ids to column &#39;nbs_2&#39; ... 
Finished. </code></pre>
<p>The default behavior is to find all spots which are neighbors with the selected id but ignoring its label, therefore it will simply be called nbs_12 as in “neighbors to 12”.</p>
<pre class="r"><code>FeatureOverlay(se, features = &quot;nbs_2&quot;, ncols.samples = 2, sampleids = 1:2, cols = c(&quot;red&quot;, &quot;black&quot;, &quot;lightgray&quot;), pt.size = 1, dark.theme = T)</code></pre>
<p><img src="figure/spatial_features.Rmd/unnamed-chunk-6-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>You can also keep all spots from the id group by setting keep.within.id = TRUE.</p>
<pre class="r"><code>se &lt;- SetIdent(se, value = &quot;seurat_clusters&quot;)
se &lt;- RegionNeighbours(se, id = 2, keep.within.id = T, verbose = TRUE)</code></pre>
<pre><code>Creating Connected Netork using KNN ... 
Found 4763 neighbours for id 2 ... 
Naming all neighbours nbs_2 ... 
Saving neighbour ids to column &#39;nbs_2&#39; ... 
Finished. </code></pre>
<pre class="r"><code>FeatureOverlay(se, features = &quot;nbs_2&quot;, ncols.samples = 2, sampleids = 1:2, cols = c(&quot;red&quot;, &quot;black&quot;, &quot;lightgray&quot;), pt.size = 1, dark.theme = T)</code></pre>
<p><img src="figure/spatial_features.Rmd/unnamed-chunk-8-1.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>se &lt;- SetIdent(se, value = &quot;nbs_2&quot;)
nbs_2.markers &lt;- FindMarkers(se, ident.1 = &quot;2&quot;, ident.2 = &quot;nbs_2&quot;)
nbs_2.markers$gene &lt;- rownames(nbs_2.markers)
se.subset &lt;- SubsetSTData(se, spots = rownames(subset(se[[]], nbs_2 %in% c(&quot;2&quot;, &quot;nbs_2&quot;))))
sorted.marks &lt;- nbs_2.markers %&gt;% top_n(n = 40, wt = abs(avg_logFC))
sorted.marks &lt;- sorted.marks[order(sorted.marks$avg_logFC, decreasing = T), ]
DoHeatmap(se.subset, features = sorted.marks$gene, group.colors = c(&quot;black&quot;, &quot;lightgray&quot;), disp.min = -2, disp.max = 2)+ DarkTheme() </code></pre>
<p><img src="figure/spatial_features.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>And lastly, if you want to keep the labels for the neighbors you can set keep.idents = TRUE and the neighbors will be called label_nb_to_12.</p>
<pre class="r"><code>se &lt;- SetIdent(se, value = &quot;seurat_clusters&quot;)
se &lt;- RegionNeighbours(se, id = 2, keep.idents = TRUE, verbose = TRUE)</code></pre>
<pre><code>Creating Connected Netork using KNN ... 
Found 4763 neighbours for id 2 ... 
Excluding neighbours from the same group ... 
569 neighbours left ... 
Naming neighbours to id_nb_to* ... 
Saving neighbour ids to column &#39;nbs_2&#39; ... 
Finished. </code></pre>
<pre class="r"><code>FeatureOverlay(se, features = &quot;nbs_2&quot;, ncols.samples = 2, sampleids = 1:2, pt.size = 1, dark.theme = T)</code></pre>
<p><img src="figure/spatial_features.Rmd/unnamed-chunk-10-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><br></p>
 
<hr />
<p style="text-align: center;">
A work by <a href="j.bergenstrahle@scilifelab.se">Joseph Bergenstråhle</a> and <a href="ludvig.larsson@scilifelab.se">Ludvig Larsson</a>
</p>
<!-- Add icon library -->
<p><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css"></p>
<!-- Add font awesome icons -->
<p style="text-align: center;">
<a href="https://www.spatialresearch.org" class="fa fa-beer"></a>
</p>
<p> </p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.1 (2019-07-05)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363)

Matrix products: default

locale:
[1] LC_COLLATE=English_Sweden.1252  LC_CTYPE=English_Sweden.1252   
[3] LC_MONETARY=English_Sweden.1252 LC_NUMERIC=C                   
[5] LC_TIME=English_Sweden.1252    

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] dplyr_0.8.3                 magrittr_1.5               
 [3] RColorBrewer_1.1-2          STutility_0.1.0            
 [5] ggplot2_3.2.1               SingleCellExperiment_1.6.0 
 [7] SummarizedExperiment_1.14.1 DelayedArray_0.10.0        
 [9] BiocParallel_1.18.1         matrixStats_0.55.0         
[11] Biobase_2.44.0              GenomicRanges_1.36.1       
[13] GenomeInfoDb_1.20.0         IRanges_2.18.2             
[15] S4Vectors_0.22.1            BiocGenerics_0.30.0        
[17] Seurat_3.1.1.9023          

loaded via a namespace (and not attached):
  [1] reticulate_1.13         R.utils_2.9.0          
  [3] tidyselect_0.2.5        htmlwidgets_1.3        
  [5] grid_3.6.1              Rtsne_0.15             
  [7] munsell_0.5.0           codetools_0.2-16       
  [9] ica_1.0-2               future_1.15.0          
 [11] miniUI_0.1.1.1          withr_2.1.2            
 [13] colorspace_1.4-1        knitr_1.24             
 [15] ROCR_1.0-7              tensor_1.5             
 [17] gbRd_0.4-11             listenv_0.7.0          
 [19] labeling_0.3            Rdpack_0.11-0          
 [21] git2r_0.26.1            GenomeInfoDbData_1.2.1 
 [23] polyclip_1.10-0         rprojroot_1.3-2        
 [25] vctrs_0.2.0             xfun_0.9               
 [27] R6_2.4.0                doParallel_1.0.15      
 [29] rsvd_1.0.2              Morpho_2.7             
 [31] ggiraph_0.6.1           manipulateWidget_0.10.0
 [33] bitops_1.0-6            spatstat.utils_1.13-0  
 [35] assertthat_0.2.1        promises_1.0.1         
 [37] SDMTools_1.1-221.1      scales_1.0.0           
 [39] imager_0.41.2           gtable_0.3.0           
 [41] npsurv_0.4-0            globals_0.12.4         
 [43] bmp_0.3                 goftest_1.1-1          
 [45] workflowr_1.5.0         rlang_0.4.1            
 [47] zeallot_0.1.0           akima_0.6-2            
 [49] systemfonts_0.1.1       splines_3.6.1          
 [51] lazyeval_0.2.2          rgl_0.100.30           
 [53] yaml_2.2.0              reshape2_1.4.3         
 [55] abind_1.4-5             crosstalk_1.0.0        
 [57] backports_1.1.4         httpuv_1.5.2           
 [59] tools_3.6.1             gplots_3.0.1.1         
 [61] raster_3.0-7            Rvcg_0.18              
 [63] ggridges_0.5.1          Rcpp_1.0.3             
 [65] plyr_1.8.4              zlibbioc_1.30.0        
 [67] purrr_0.3.3             RCurl_1.95-4.12        
 [69] dbscan_1.1-5            rpart_4.1-15           
 [71] deldir_0.1-23           viridis_0.5.1          
 [73] pbapply_1.4-2           cowplot_1.0.0          
 [75] zoo_1.8-6               ggrepel_0.8.1          
 [77] cluster_2.1.0           colorRamps_2.3         
 [79] fs_1.3.1                RSpectra_0.15-0        
 [81] data.table_1.12.2       magick_2.2             
 [83] readbitmap_0.1.5        lmtest_0.9-37          
 [85] RANN_2.6.1              fitdistrplus_1.0-14    
 [87] shinyjs_1.0             lsei_1.2-0             
 [89] mime_0.7                evaluate_0.14          
 [91] xtable_1.8-4            jpeg_0.1-8.1           
 [93] gridExtra_2.3           compiler_3.6.1         
 [95] tibble_2.1.3            KernSmooth_2.23-15     
 [97] crayon_1.3.4            R.oo_1.22.0            
 [99] htmltools_0.3.6         mgcv_1.8-28            
[101] later_0.8.0             tiff_0.1-5             
[103] tidyr_1.0.0             RcppParallel_4.4.3     
[105] MASS_7.3-51.4           Matrix_1.2-17          
[107] R.methodsS3_1.7.1       gdata_2.18.0           
[109] metap_1.1               igraph_1.2.4.1         
[111] pkgconfig_2.0.2         sp_1.3-2               
[113] plotly_4.9.1            xml2_1.2.2             
[115] foreach_1.4.7           webshot_0.5.1          
[117] XVector_0.24.0          bibtex_0.4.2           
[119] stringr_1.4.0           digest_0.6.22          
[121] sctransform_0.2.0       RcppAnnoy_0.0.12       
[123] tsne_0.1-3              spatstat.data_1.4-0    
[125] rmarkdown_1.15          leiden_0.3.1           
[127] uwot_0.1.4              gdtools_0.2.0          
[129] shiny_1.3.2             gtools_3.8.1           
[131] lifecycle_0.1.0         nlme_3.1-141           
[133] jsonlite_1.6            viridisLite_0.3.0      
[135] pillar_1.4.2            lattice_0.20-38        
[137] httr_1.4.1              survival_2.44-1.1      
[139] glue_1.3.1              spatstat_1.61-0        
[141] png_0.1-7               iterators_1.0.12       
[143] stringi_1.4.3           caTools_1.17.1.2       
[145] irlba_2.3.3             future.apply_1.3.0     
[147] ape_5.3                </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
