<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="" />


<title>Spatial Features</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">STUtility</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="installation.html">Installation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Getting started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="getting_started.html">Load data</a>
    </li>
    <li>
      <a href="Quality_Control.html">Quality Control</a>
    </li>
    <li>
      <a href="image_processing.html">Image processing</a>
    </li>
    <li>
      <a href="Manual_annotation.html">Manual annotation</a>
    </li>
    <li>
      <a href="normalization.html">Normalization</a>
    </li>
    <li>
      <a href="spatial_features.html">Spatial features</a>
    </li>
    <li>
      <a href="Neighborhood.html">Neighborhood Analysis</a>
    </li>
    <li>
      <a href="3D_visualization.html">3D visualization</a>
    </li>
    <li>
      <a href="Split_Data.html">Split Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/ludvigla/STUtility_web_site">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Spatial Features</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-05-05
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>STUtility_web_site/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20191031code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20191031)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20191031code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20191031)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomludviglaSTUtilitywebsitetreebc60b884b5d96ce650a0bff1c18abbc66bc64662targetblankbc60b88a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/ludvigla/STUtility_web_site/tree/bc60b884b5d96ce650a0bff1c18abbc66bc64662" target="_blank">bc60b88</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomludviglaSTUtilitywebsitetreebc60b884b5d96ce650a0bff1c18abbc66bc64662targetblankbc60b88a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/ludvigla/STUtility_web_site/tree/bc60b884b5d96ce650a0bff1c18abbc66bc64662" target="_blank">bc60b88</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/manual_annotation.png
    Ignored:    pre_data/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with <code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<style type="text/css">
div.main-container {
  background-color: #FFFFFF !important;
  max-width: 1800px;
  margin-left: auto;
  margin-right: auto;
}
</style>
<style>
#TOC {
  background: url("https://www.spatialresearch.org/wp-content/uploads/2019/09/str-logo-spatial_research_3@2x.png");
  background-size: contain;
  padding-top: 100px !important;
  background-repeat: no-repeat;
  width: 500px;
  color: black;
  border-color: #FFFFFF !important;
}
</style>
<style> code, pre{
  background-color: #F8F8F8 !important;
  color: black !important;
}
</style>
<style>
.watch-out { background-color: #F0F0F0; border: 1px #F0F0F0; }
<style>
<style> 
body {
  color: black;
  font-size: 20pt;
}
</style>
<div id="spatial-auto-correlation" class="section level2">
<h2>Spatial Auto-correlation</h2>
<hr />
<p>STutility includes a method for finding genes with spatially conserved patterns across the tissue. The ranking method makes use neighborhood networks to compute the spatial lag for each gene, here defined as the summed expression of that gene in neighboring spots. Each gene is then ranked by the correlation between the lag vector and the original expression vector. The output is a <code>data.frame</code> with gene names ranked by decreasing spatial auto-correlation.</p>
<p>This method is partly inspired by work from the <a href="http://spatial.rc.fas.harvard.edu/spatialgiotto/giotto.html">Giotto team</a> and we reccomend you to check out their R package Giotto and the related publication; [“Giotto: a toolbox for integrative analysis and visualization of spatial expression data”](<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02286-2" class="uri">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02286-2</a>.</p>
<p><br></p>
<pre class="r"><code>library(spdep)
spatgenes &lt;- CorSpatialGenes(se)</code></pre>
<p>By default, the saptial-auto-correlation scores are only calculated for the variable genes in the Seurat object, here we have 3000. Among the top most variable features in our Seurat object, we find genes coding for hemoglobin; “Hbb-bs” “Hba-a1” “Hba-a2”. These are typically expressed in blood vessels which are more randomly distributed across the tissue compared to larger tissue structures. Knowing the spatial auto-correlation can therefore be useful to distinguish genes expressed in larger tissue compartments. One way to make use of this is to restrict the selection of features used for dimensionality reduction and clustering to include ony gene that are highly variable and spatially auto-correlated, and that way avoid clustering based on structures such as blood vessels.</p>
<pre class="r"><code>head(VariableFeatures(se))</code></pre>
<pre><code>[1] &quot;Hbb-bs&quot; &quot;Hba-a1&quot; &quot;Hba-a2&quot; &quot;Plp1&quot;   &quot;Mbp&quot;    &quot;Ptgds&quot; </code></pre>
<pre class="r"><code>spatgenes[c(&quot;Hbb-bs&quot;, &quot;Hba-a1&quot;, &quot;Hba-a2&quot;), ]</code></pre>
<pre><code>         gene       cor
Hbb-bs Hbb-bs 0.3864246
Hba-a1 Hba-a1 0.3285132
Hba-a2 Hba-a2 0.3152952</code></pre>
<p>Let’s plot some of the genes with highest spatial auto-correlation.</p>
<pre class="r"><code>head(spatgenes) %&gt;%
  kbl() %&gt;%
  kable_styling()</code></pre>
<table class="table" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
gene
</th>
<th style="text-align:right;">
cor
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Mbp
</td>
<td style="text-align:left;">
Mbp
</td>
<td style="text-align:right;">
0.9203725
</td>
</tr>
<tr>
<td style="text-align:left;">
Camk2n1
</td>
<td style="text-align:left;">
Camk2n1
</td>
<td style="text-align:right;">
0.9054777
</td>
</tr>
<tr>
<td style="text-align:left;">
Slc6a3
</td>
<td style="text-align:left;">
Slc6a3
</td>
<td style="text-align:right;">
0.8917767
</td>
</tr>
<tr>
<td style="text-align:left;">
Th
</td>
<td style="text-align:left;">
Th
</td>
<td style="text-align:right;">
0.8762578
</td>
</tr>
<tr>
<td style="text-align:left;">
Nrgn
</td>
<td style="text-align:left;">
Nrgn
</td>
<td style="text-align:right;">
0.8744128
</td>
</tr>
<tr>
<td style="text-align:left;">
Tmsb4x
</td>
<td style="text-align:left;">
Tmsb4x
</td>
<td style="text-align:right;">
0.8715944
</td>
</tr>
</tbody>
</table>
<pre class="r"><code>FeatureOverlay(se, features = c(&quot;Mbp&quot;, &quot;Camk2n1&quot;, &quot;Slc6a3&quot;, &quot;Th&quot;), 
              sampleids = 1,
              cols = c(&quot;lightgray&quot;, &quot;mistyrose&quot;, &quot;red&quot;, &quot;darkred&quot;, &quot;black&quot;),
              pt.size = 1.5, 
              add.alpha = TRUE,
              ncol = 2)</code></pre>
<p><img src="figure/Spatial_Features.Rmd/unnamed-chunk-2-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="matrix-factorization" class="section level2">
<h2>Matrix factorization</h2>
<hr />
<p>The strength of untargeted whole transcriptome capture is the ability to perform unsupervised analysis and the ability to find spatial gene expression patterns. We’ve found good use of using non-negative matrix factorization (NNMF or NMF) to find underlying patterns of transcriptomic profiles. This factor analysis, along with various dimensionality reduction techniques, can all be ran via “RunXXX()”, where X = the method of choice, e.g.: <br></p>
<pre class="r"><code>se &lt;- RunNMF(se, nfactors = 40) # Specificy nfactors to choose the number of factors, default=20.</code></pre>
<p><br> While <code>RunNMF()</code> is an STutility add-on, others are supported via Seurat (<code>RunPCA()</code>, <code>RunTSNE</code>, <code>RunICA()</code>, <code>runUMAP()</code> ) and for all of them, the output are stored in the Seurat object.</p>
<p>We can then plot a variable number of dimensions across the samples using <code>ST.DimPlot</code> or as an overlay using <code>DimOverlay</code>. These two functions are similar to the <code>ST.FeaturePlot</code> and <code>FeatureOverlay</code> but have been adapted to specifically draw dimensionality reduction vectors instead of features.</p>
<p>NOTE: by default, the colorscale of dimensionality reduction vectors will be centered at 0. If we have a dimensionality reduction vector x this means that the range of colors will go from -max(abs(x)) to max(abs(x)). This behaviour is typically desired when plotting e.g. PCA vectors, but for NMF all values are strictly positive so you can disable this centering by setting <code>center.zero = FALSE</code>. <br></p>
<pre class="r"><code>cscale &lt;- c(&quot;lightgray&quot;, &quot;mistyrose&quot;, &quot;red&quot;, &quot;darkred&quot;, &quot;black&quot;)

ST.DimPlot(se, 
           dims = 1:10,
           ncol = 2, # Sets the number of columns at dimensions level
           grid.ncol = 2, # Sets the number of columns at sample level
           reduction = &quot;NMF&quot;, 
           pt.size = 1, 
           center.zero = F, 
           cols = cscale, 
           show.sb = FALSE)</code></pre>
<p><img src="figure/Spatial_Features.Rmd/dimplot-1.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ST.DimPlot(se, 
           dims = 11:20,
           ncol = 2, 
           grid.ncol = 2, 
           reduction = &quot;NMF&quot;, 
           pt.size = 1, 
           center.zero = F, 
           cols = cscale, 
           show.sb = FALSE)</code></pre>
<p><img src="figure/Spatial_Features.Rmd/dimplot-2.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ST.DimPlot(se, 
           dims = 21:30,
           ncol = 2, 
           grid.ncol = 2, 
           reduction = &quot;NMF&quot;, 
           pt.size = 1, 
           center.zero = F, 
           cols = cscale, 
           show.sb = FALSE)</code></pre>
<p><img src="figure/Spatial_Features.Rmd/dimplot-3.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ST.DimPlot(se, 
           dims = 31:40,
           ncol = 2, 
           grid.ncol = 2, 
           reduction = &quot;NMF&quot;, 
           pt.size = 1, 
           center.zero = F, 
           cols = cscale, 
           show.sb = FALSE)</code></pre>
<p><img src="figure/Spatial_Features.Rmd/dimplot-4.png" width="100%" style="display: block; margin: auto;" /> <br> We can also print a summary of the genes that contribute most to the dimensionality reduction vectors.</p>
<p>For NMF output which is not centered at 0 looking at the “negative” side of the distribution doesn’t really add any valuable information, instead you can aget a barplot summarizing the top most contributing genes using <code>FactorGeneLoadingPlot</code>. <br></p>
<pre class="r"><code>print(se[[&quot;NMF&quot;]])</code></pre>
<pre><code>factor_ 1 
Positive:  Pvalb, Gad1, Spp1, Slc32a1, Six3, Sparc, Lsm6, Ndufb8, Nsg1, Agt 
       Enho, Ldhb, Gabra1, Cst3, Cox6c, Ndrg2, Cox5a, Ckb, Lgi2, Nefh 
Negative:  Rgs20, C1qtnf2, Lyn, Nid2, St6galnac2, Rbm20, Pdzd2, Mndal, Emx2, Sp9 
       Gcnt2, Gm19410, Irak2, Tbx2, Gm11627, Nfatc1, Bambi, Sorcs1, Pam, Arhgap45 
factor_ 2 
Positive:  Vamp1, Pvalb, Bend6, Nat8l, Cox5a, Pcp4l1, Stmn3, Nefm, Syt2, Snrpn 
       Pcsk1n, Ndufa4, Ctxn3, Slc17a6, Cend1, Kcnab3, Cox4i1, Mdh1, Scn1b, Camk2n2 
Negative:  Plxnc1, Cdhr1, C1qtnf2, Evc, Klf11, Nid2, Ly6g6f, Hist1h4h, Ppp1r18, St6galnac2 
       Fxyd2, Blnk, Fam89a, Dock8, Emx2, Cyp20a1, Sp9, Ginm1, Acacb, Gm10561 
factor_ 3 
Positive:  mt-Nd3, Rps29, Rps21, Rpl39, Rplp1, mt-Co3, Cnot3, Rpl37, Rps19, Rps27 
       mt-Nd5, Rpl35a, Rps28, mt-Nd4l, Rpl37a, Tatdn1, Rpl26, Nnat, Uba52, mt-Co1 
Negative:  Otx1, Rgs20, Evc, Pkd2, Nid2, Ppp1r18, St6galnac2, Tnxb, Atf4, Rbm20 
       Slc25a45, Pdzd2, Mndal, Amt, Blnk, Agtrap, Grik1, Fam89a, Dock8, Pcdh18 
factor_ 4 
Positive:  Th, Slc6a3, Slc18a2, Ddc, Sncg, Slc10a4, Ret, Dlk1, En1, Chrna6 
       Drd2, Cplx1, Sv2c, Aldh1a1, Gch1, Uchl1, Gap43, Chrnb3, Tagln3, Foxa1 
Negative:  Cdhr1, Evc, Nid2, St6galnac2, Slc25a45, Blnk, C530008M17Rik, Dock8, Emx2, Sp9 
       Bin2, Wnt5b, Nfatc1, Bambi, Sorcs1, Phactr4, Masp1, Rapgef3, Kdr, Epha8 
factor_ 5 
Positive:  Spink8, Tmsb4x, Fibcd1, Hpca, Fkbp1a, Itpka, Cnih2, Calm2, Tspan13, Lefty1 
       Crym, Prnp, Dynll1, Rprml, Cpne6, Arpc5, Mpped1, Cpne7, Neurod6, Ociad2 
Negative:  Cdh9, Otx1, Spsb1, Cdhr1, Rgs20, Nid2, Ly6g6f, Ppp1r18, Rbm20, Slc25a45 
       Pdzd2, Mndal, Atp2a3, Grik1, Fam89a, Cux1, Tex9, Dock8, Msi1, Sp9 </code></pre>
<pre class="r"><code>FactorGeneLoadingPlot(se, factor = 1)</code></pre>
<p><img src="figure/Spatial_Features.Rmd/project_dim-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="clustering" class="section level2">
<h2>Clustering</h2>
<hr />
<p>Clustering is a standard procedure in genomic analysis, and the methods for doing so are numerous. Here we demonstrate an example where we use the result of the factor analysis the previous section. Going through the list of factors (e.g. via <code>ST:DimPlot(se, dims = [dims you want to look at])</code>), we can notice dimensions that are “spatially active”, i.e. that seems to confer a spatial pattern along their axis. We can extract these dimensions and use as input to e.g. clustering functions. Here, we use all dimensions from the NMF and construct a Shared Nearest Neighbor (SSN) Graph. <br></p>
<pre class="r"><code>se &lt;- FindNeighbors(object = se, verbose = FALSE, reduction = &quot;NMF&quot;, dims = 1:40)</code></pre>
<p><br> Followed by clustering using a modularity optimizer <br></p>
<pre class="r"><code>se &lt;- FindClusters(object = se, verbose = FALSE)</code></pre>
<p><br> And plotting of the clusters spatially <br></p>
<pre class="r"><code>library(RColorBrewer)
n &lt;- 19
qual_col_pals = brewer.pal.info[brewer.pal.info$category == &#39;qual&#39;,]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))

ST.FeaturePlot(object = se, features = &quot;seurat_clusters&quot;, cols = col_vector, pt.size = 1, ncol = 2)</code></pre>
<p><img src="figure/Spatial_Features.Rmd/plot_clusters-1.png" width="100%" style="display: block; margin: auto;" /> <br> If you think that the distribution of clusters gets too cluttered, you can also split the view so that only one cluster at the time gets colored, just note that you can only do this for one section at the time (set <code>ìndex</code>). <br></p>
<pre class="r"><code>ST.FeaturePlot(object = se, features = &quot;seurat_clusters&quot;, pt.size = 1, split.labels = T, indices = 1, show.sb = FALSE, ncol = 5)</code></pre>
<p><img src="figure/Spatial_Features.Rmd/plot_clusters_split-1.png" width="100%" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ST.FeaturePlot(object = se, features = &quot;seurat_clusters&quot;, pt.size = 1, split.labels = T, indices = 2, show.sb = FALSE, ncol = 5)</code></pre>
<p><img src="figure/Spatial_Features.Rmd/plot_clusters_split-2.png" width="100%" style="display: block; margin: auto;" /> <br></p>
</div>
<div id="most-variable-features" class="section level2">
<h2>Most variable features</h2>
<hr />
<p>We can take a specific look at some of the most variable features defined during the normalization step. <br></p>
<pre class="r"><code>head(se@assays$SCT@var.features, 20)</code></pre>
<pre><code> [1] &quot;Hbb-bs&quot;  &quot;Hba-a1&quot;  &quot;Hba-a2&quot;  &quot;Plp1&quot;    &quot;Mbp&quot;     &quot;Ptgds&quot;   &quot;Hbb-bt&quot; 
 [8] &quot;Slc6a3&quot;  &quot;Sst&quot;     &quot;Th&quot;      &quot;Ddc&quot;     &quot;Npy&quot;     &quot;Slc18a2&quot; &quot;Mobp&quot;   
[15] &quot;Nrgn&quot;    &quot;Mal&quot;     &quot;Pcp4&quot;    &quot;Prkcd&quot;   &quot;Apod&quot;    &quot;Myoc&quot;   </code></pre>
<pre class="r"><code>top &lt;- se@assays$SCT@var.features

fts &lt;- c(&quot;Th&quot;, &quot;Mbp&quot;, &quot;Nrgn&quot;)
for (ftr in fts) {
  p &lt;- FeatureOverlay(se, 
                  features = ftr, 
                  sampleids = 1:2,
                  cols = c(&quot;lightgray&quot;, &quot;mistyrose&quot;, &quot;red&quot;, &quot;darkred&quot;, &quot;black&quot;),
                  pt.size = 1.5, 
                  pt.alpha = 0.5, 
                 ncols = 2)
  print(p)
}</code></pre>
<p><img src="figure/Spatial_Features.Rmd/variable_features-1.png" width="100%" style="display: block; margin: auto;" /><img src="figure/Spatial_Features.Rmd/variable_features-2.png" width="100%" style="display: block; margin: auto;" /><img src="figure/Spatial_Features.Rmd/variable_features-3.png" width="100%" style="display: block; margin: auto;" /> <br></p>
</div>
<div id="spatial-vs.umap-visualziation" class="section level2">
<h2>Spatial vs. UMAP visualziation</h2>
<hr />
<p>Another useful feature is that you can now compare the spatial distribution of a gene with the typical “graph embeddings” s.a. UMAP and t-SNE. <br></p>
<pre class="r"><code># Run UMAP
se &lt;- RunUMAP(se, reduction = &quot;NMF&quot;, dims = 1:40, n.neighbors = 10)</code></pre>
<pre class="r"><code># Define colors for heatmap
heatmap.colors &lt;- c(&quot;lightgray&quot;, &quot;mistyrose&quot;, &quot;red&quot;, &quot;darkred&quot;, &quot;black&quot;)
fts &lt;- c(&quot;Prkcd&quot;, &quot;Opalin&quot;, &quot;Lamp5&quot;)

# plot transformed features expression on UMAP embedding
p.fts &lt;- lapply(fts, function(ftr) {
  FeaturePlot(se, features = ftr, reduction = &quot;umap&quot;, order = TRUE, cols = heatmap.colors)
})

# plot transformed features expression on Visium coordinates
p3 &lt;- ST.FeaturePlot(se, features = fts, ncol = 2, grid.ncol = 1, cols = heatmap.colors, pt.size = 1, show.sb = FALSE)

# Construct final plot
cowplot::plot_grid(cowplot::plot_grid(plotlist = p.fts, ncol = 1), p3, ncol = 2, rel_widths = c(1, 1.3))</code></pre>
<p><img src="figure/Spatial_Features.Rmd/embedding_vs_ST-1.png" width="1152" style="display: block; margin: auto;" /></p>
</div>
<div id="rgb-dimensionality-reduction-plots" class="section level2">
<h2>RGB dimensionality reduction plots</h2>
<hr />
<p>One approach to visualize the result of dimensionality reduction is to use the first three dimensions and transform the values into RGB color space. This 3 dimensional space can then be utilized for spatial visualization. We demonstrate the technique with UMAP, using our factors as input: <br></p>
<pre class="r"><code>se &lt;- RunUMAP(object = se, dims = 1:40, verbose = FALSE, n.components = 3, reduction = &quot;NMF&quot;, reduction.name = &quot;umap.3d&quot;)</code></pre>
<p><br> We use the first three dimensions for plotting: <br></p>
<pre class="r"><code>ST.DimPlot(object = se, dims = 1:3, reduction = &quot;umap.3d&quot;, blend = T, pt.size = 1.8)</code></pre>
<p><img src="figure/Spatial_Features.Rmd/UMAP_blend-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p><br></p>
</div>
<div id="dea" class="section level2">
<h2>DEA</h2>
<hr />
<p>Lets try this out by an example. Looking at , lets say we are interested in cluster 19, and want to answer - “Which genes are significantly up-regulated in this region compared to the overall expression across the tissue?”</p>
<pre class="r"><code>markers &lt;- FindMarkers(se, ident.1 = &quot;19&quot;)
head(markers) %&gt;%
  kbl() %&gt;%
  kable_styling()</code></pre>
<table class="table" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
p_val
</th>
<th style="text-align:right;">
avg_logFC
</th>
<th style="text-align:right;">
pct.1
</th>
<th style="text-align:right;">
pct.2
</th>
<th style="text-align:right;">
p_val_adj
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Dsp
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.2427413
</td>
<td style="text-align:right;">
0.818
</td>
<td style="text-align:right;">
0.015
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Lct
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.9157160
</td>
<td style="text-align:right;">
0.768
</td>
<td style="text-align:right;">
0.021
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Tdo2
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0.7143539
</td>
<td style="text-align:right;">
0.707
</td>
<td style="text-align:right;">
0.026
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Capn3
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.1263876
</td>
<td style="text-align:right;">
0.859
</td>
<td style="text-align:right;">
0.084
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Prox1
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.1988020
</td>
<td style="text-align:right;">
0.919
</td>
<td style="text-align:right;">
0.117
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
C1ql2
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
1.7382051
</td>
<td style="text-align:right;">
1.000
</td>
<td style="text-align:right;">
0.169
</td>
<td style="text-align:right;">
0
</td>
</tr>
</tbody>
</table>
<p>Note that the clusters were already set as the Seurat objects levels. Type <code>levels(se)</code> to see the current levels of your object. If other clusters, annotations etc are of interest, set this before by specifying <code>Idents(se) &lt;-</code></p>
<p>Note also, if we are interested in comparing two levels against each other, and not just “one against the rest”, we simply add a <code>ident.2 =</code> parameter to the above.</p>
<pre class="r"><code>FeatureOverlay(se, features = &quot;Dsp&quot;, 
              sampleids = 1:2,
              cols = c(&quot;lightgray&quot;, &quot;mistyrose&quot;, &quot;red&quot;, &quot;darkred&quot;, &quot;black&quot;),
              pt.size = 1.5, 
              pt.alpha = 0.5,
              ncol = 2)</code></pre>
<p><img src="figure/Spatial_Features.Rmd/unnamed-chunk-3-1.png" width="100%" style="display: block; margin: auto;" /> <br></p>
 
<hr />
<p style="text-align: center;">
A work by <a href="mailto: j.bergenstrahle@scilifelab.se">Joseph Bergenstråhle</a> and <a href="mailto: ludvig.larsson@scilifelab.se">Ludvig Larsson</a>
</p>
<!-- Add icon library -->
<p><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css"></p>
<!-- Add font awesome icons -->
<p style="text-align: center;">
<a href="https://www.spatialresearch.org" class="fa fa-beer"></a>
</p>
<p> </p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.5 (2021-03-31)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] RColorBrewer_1.1-2 magrittr_2.0.1     kableExtra_1.3.4   STutility_0.1.0   
[5] ggplot2_3.3.3      Seurat_3.1.5      

loaded via a namespace (and not attached):
  [1] uuid_0.1-4              backports_1.1.6         workflowr_1.6.2        
  [4] systemfonts_0.2.1       plyr_1.8.6              igraph_1.2.6           
  [7] lazyeval_0.2.2          sp_1.4-1                splines_4.0.5          
 [10] crosstalk_1.1.1         listenv_0.8.0           digest_0.6.27          
 [13] foreach_1.5.0           htmltools_0.5.1.1       viridis_0.5.1          
 [16] magick_2.3              tiff_0.1-5              gdata_2.18.0           
 [19] fansi_0.4.2             cluster_2.1.1           doParallel_1.0.15      
 [22] ROCR_1.0-11             limma_3.44.1            globals_0.14.0         
 [25] gmodels_2.18.1          svglite_1.2.3           jpeg_0.1-8.1           
 [28] colorspace_2.0-0        rvest_0.3.6             blob_1.2.1             
 [31] ggrepel_0.9.1           xfun_0.13               dplyr_1.0.5            
 [34] crayon_1.4.1            jsonlite_1.7.2          zeallot_0.1.0          
 [37] survival_3.2-10         zoo_1.8-9               iterators_1.0.12       
 [40] ape_5.4-1               glue_1.4.2              gtable_0.3.0           
 [43] webshot_0.5.2           leiden_0.3.7            future.apply_1.7.0     
 [46] scales_1.1.1            DBI_1.1.0               miniUI_0.1.1.1         
 [49] Rcpp_1.0.6              viridisLite_0.3.0       xtable_1.8-4           
 [52] spData_0.3.5            units_0.6-6             reticulate_1.18        
 [55] spdep_1.1-3             rsvd_1.0.3              akima_0.6-2            
 [58] tsne_0.1-3              htmlwidgets_1.5.3       httr_1.4.2             
 [61] ellipsis_0.3.1          ica_1.0-2               farver_2.1.0           
 [64] pkgconfig_2.0.3         uwot_0.1.10             deldir_0.1-25          
 [67] utf8_1.2.1              labeling_0.4.2          tidyselect_1.1.0       
 [70] rlang_0.4.10            manipulateWidget_0.10.1 reshape2_1.4.4         
 [73] later_1.1.0.1           munsell_0.5.0           tools_4.0.5            
 [76] dbscan_1.1-5            generics_0.1.0          ggridges_0.5.3         
 [79] evaluate_0.14           stringr_1.4.0           fastmap_1.0.1          
 [82] yaml_2.2.1              knitr_1.28              fs_1.5.0               
 [85] fitdistrplus_1.1-3      rgl_0.100.54            purrr_0.3.4            
 [88] RANN_2.6.1              readbitmap_0.1.5        pbapply_1.4-3          
 [91] future_1.21.0           nlme_3.1-152            mime_0.10              
 [94] ggiraph_0.7.7           xml2_1.3.2              rstudioapi_0.13        
 [97] compiler_4.0.5          plotly_4.9.3            png_0.1-7              
[100] e1071_1.7-3             Morpho_2.8              tibble_3.1.0           
[103] stringi_1.5.3           highr_0.8               gdtools_0.2.2          
[106] lattice_0.20-41         Matrix_1.3-2            classInt_0.4-3         
[109] shinyjs_1.1             vctrs_0.3.7             pillar_1.5.1           
[112] LearnBayes_2.15.1       lifecycle_1.0.0         lmtest_0.9-38          
[115] RcppAnnoy_0.0.18        data.table_1.14.0       cowplot_1.1.1          
[118] irlba_2.3.3             Rvcg_0.19.1             raster_3.1-5           
[121] httpuv_1.5.2            patchwork_1.1.1         colorRamps_2.3         
[124] R6_2.5.0                imager_0.42.1           promises_1.2.0.1       
[127] KernSmooth_2.23-18      gridExtra_2.3           bmp_0.3                
[130] parallelly_1.24.0       codetools_0.2-18        gtools_3.8.2           
[133] boot_1.3-27             MASS_7.3-53.1           assertthat_0.2.1       
[136] rprojroot_1.3-2         withr_2.4.1             sctransform_0.2.1      
[139] expm_0.999-4            parallel_4.0.5          grid_4.0.5             
[142] tidyr_1.1.3             coda_0.19-3             class_7.3-18           
[145] rmarkdown_2.1           Rtsne_0.15              git2r_0.27.1           
[148] sf_0.9-7                shiny_1.4.0.2          </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
