---
title: "About"
author:
- Joseph Bergenstråhle, SciLifeLab, Royal Institute of Technology (KTH)
- Ludvig Larsson, SciLifeLab, Royal Institute of Technology (KTH)
date: ''
---

<style type="text/css">
div.main-container {
  background-color: #FFFFFF !important;
  max-width: 1800px;
  margin-left: auto;
  margin-right: auto;
}
</style>

<style>
#TOC {
  background: url("https://www.spatialresearch.org/wp-content/uploads/2019/09/str-logo-spatial_research_3@2x.png");
  background-size: contain;
  padding-top: 100px !important;
  background-repeat: no-repeat;
  width: 500px;
  color: black;
  border-color: #FFFFFF !important;
}
</style>

<style> code, pre{
  background-color: #F8F8F8 !important;
  color: black !important;
}
</style>

<style> .watch-out {
  background-color: #F0F0F0;
  border: 1px #F0F0F0;
}
<style>

<style> 
body {
  color: black;
  font-size: 20pt;
}
</style>



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, autodep = FALSE)
```

# Background

STutility R-package is an effort to create a user-friendly infrastructure for analysis of spatial transcriptomcis data. The tool is built to work on top of [Seurat](https://github.com/satijalab/seurat), which is a popular single-cell genomics analysis toolkit. 

### Spatial Transcriptomics (ST)

---

Spatial Transcriptomics is a method that allows visualization and quantitative analysis of the transcriptome in thin tissue sections by combining gene expression data and microscopy based image data. The invention was first presented [in science](https://science.sciencemag.org/content/353/6294/78) in 2016 and has since been commercialized by 10x Genomics. The method has received increasing attention and is currently the basis of several national and international collaborations and spatially resolved transcriptomics was recently awarded [method of the year in 2020](https://www.nature.com/articles/s41592-020-01038-7). 

The schematic below gives a brief overview of the concept.
An introductory animation is available on our website: http://www.spatialresearch.org/

![Schematic Spatial Transcriptomics](assets/st_method_1.PNG)

The first generation array featured 1000 capture-spots, each spot with a diameter of 100 µm. Accordingly, we refer this first generation array as the "1k" array in this tutorial and package parameters. Since the first generation array, the method has been developed to feature 2000 capture-spots ("2k" array) and 5000 spots (10x Visium).

### 10X Visium

---

This array features 5000 capture-spots, each with a diameter of 55µm printed in a hexagonal pattern. The array is [available](https://www.10xgenomics.com/spatial-transcriptomics/) for commercial use.


# Notes about the tool


### "1k" and "2k" ST arrays

The gene expression data consists of a count matrix with genes in rows and "capture-spots" in columns. Each spot represents a small area on an ST array from which the captured transcripts have been barcoded with a unique sequence. The unique barcode makes it possible to map the transcripts onto a spatial position on the tissue section and would be equivalent to a cell specific barcode in scRNA-seq data but can tag a mixture of transcripts from multiple cells. The spatial position of a spot is an (x, y) coordinate that defines the centroid of the spot area. These spatial coordinates are stored in the spot ids (column names) and allows us to visualize gene expression (and other spot features) in the array grid system. However, if you want to overlay a visualization on top the HE image you want to make sure that the spot coordinates are exact in relation to morphological features of the image. When the spots are printed onto the ST array surface, they will sometimes deviate from the (x, y) coordinates given by the spot ids and should therefore be adjusted. In addition to the spot adjustment, you will also need to label the spots that are located directly under the tissue. Spot adjustment and selection can be done automatically using our [ST spot detector](https://github.com/SpatialTranscriptomicsResearch/st_spot_detector) web tool which outputs a table of adjusted coordinates and labels for the spots under tissue.

### 10X Visium arrays

10X Genomics provide the [SpaceRanger](https://support.10xgenomics.com/spatial-gene-expression/software/downloads/latest#spacerangertab) command line tool to process raw data from an NGS instrument. With this tool, the spots under the tissue are automatically selected using a tissue detection algorithm so there's usually no need to any manual alignment steps. However, sometimes the tissue detection fails in which case it is possible to do the alignment and spot selection manually with the Loupe browser desktop application.

### Seurat workflow

The STutility package is designed to work on top of the popular single-cell genomics analysis toolkit [Seurat](https://satijalab.org/seurat/). Although a spot represents an averaged expression profile from multiple cells, as opposed to a single-cell profile, the characteristics of the ST data resembles that of scRNAseq to a large extent with regards to count distributions and sparsity. Many of the analyses that have been developed for scRNA-seq data can therefore be directly applied to ST data as well. 

### Naming conventions

For users familiar with the Seurat workflow, there are two commonly used plotting functions available to draw feature expression in a 2D-embedding such as UMAP or t-SNE, namely `FeaturePlot()` and `DimPlot`. These two plotting functions are available in STutility where "ST." is prefixed to the function name, but will instead draw feature expression in the spatial coordinate system.

&nbsp;
<hr />
<p style="text-align: center;">A work by <a href="mailto: j.bergenstrahle@scilifelab.se">Joseph Bergenstråhle</a> and <a href="mailto: ludvig.larsson@scilifelab.se">Ludvig Larsson</a></p>

<!-- Add icon library -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">

<!-- Add font awesome icons -->
<p style="text-align: center;">
    <a href="https://www.spatialresearch.org" class="fa fa-beer"></a>
</p>

&nbsp;
