---
title: "Load data"
date: ''
---

<style type="text/css">
div.main-container {
  background-color: #FFFFFF !important;
  max-width: 1800px;
  margin-left: auto;
  margin-right: auto;
}
</style>

<style>
#TOC {
  background: url("https://www.spatialresearch.org/wp-content/uploads/2019/09/str-logo-spatial_research_3@2x.png");
  background-size: contain;
  padding-top: 100px !important;
  background-repeat: no-repeat;
  width: 500px;
  color: black;
  border-color: #FFFFFF !important;
}
</style>

<style> code, pre{
  background-color: #F8F8F8 !important;
  color: black !important;
}
</style>

<style> .watch-out {
  background-color: #F0F0F0;
  border: 1px #F0F0F0;
}
<style>

<style> 
body {
  color: black;
  font-size: 20pt;
}
</style>


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, autodep = TRUE)
```

First you need to load the library into your R session.

```{r load_lib}
library(STutility)
```
```{r load_lib_hidden, include=FALSE}
library(kableExtra)
library(magrittr)
```

## 10X Visium platform
***

### Input files
***

10X Visium data output is produced with the [spaceranger](https://support.10xgenomics.com/spatial-gene-expression/software/visualization/latest/installation) command line tool from raw fastq files. The output includes a number of files, and the ones that needs to be imported into R for STutility are the following:

1. "filtered_feature_bc_matrix.h5" or "raw_feature_bc_matrix.h5" : Gene expression matrices in .h5 format containing the raw UMI counts for each spot and each gene. The "filtered_feature_bc_matrix.h5" matrix contains spots that are localized under the tissue while the "raw_feature_bc_matrix.h5" matrix contain all spots from the entire capture area. NOTE: if you want to include spots outside of the tissue, you need to set `disable.subset = TRUE` when running `InputFromTable`.
2. "tissue_positions_list.csv" : contains capture-spot barcode and capture-spot coordinates.
3. "tissue_hires_image.png" : H&E image in PNG format with a width of 2000 pixels.
4. "scalefactors_json.json" : This file contains scaling factors subject to the H&E images of different resolutions. E.g. "tissue_hires_scalef": 0.1, means that the pixel coordinates in the "tissue_positions_list.csv" table should be scaled by a factor of 0.1 to match the size of the "hires_image.png" file. 

To use the full range of functions within STutility, all four files are needed for each sample. However, all data analysis steps that do not involve the H&E image can be performed with only the count file as input. To read in the 10x Visium .h5 files, the package `hdf5r` needs to be installed: `BiocManager::install("hdf5r")`.

---

To follow along this tutorial with a test data set, go to the [10x Dataset repo](https://support.10xgenomics.com/spatial-gene-expression/datasets) and download the following two files:

* Feature / cell matrix HDF5 (filtered)
* Spatial imaging data (.zip)
    * tissue_hires_image
    * tissue_positions_list
    * scalefactors_json

The .zip file contains the H&E image (in two resolution formats; "tissue_lowres_image" and "tissue_hires_image"), the "tissue_positions_list" with pixel coordinates for the orginial TIF image and the "scalefactors_json.json" that contains the scalefactors used to derive the pixel cooridinates for the hires images. There are three alternatives to handle the scaling of pixel coordinates. Either, you manualy open the JSON file and note the scalefactor and state these numbers in a column in the infoTable named "scaleVisium" (see below). Or, you add a column named "json" with paths to the "scalefactors_json.json" files. A third option is to manually input the values to the function InputFromTable (see `?InputFromTable`). If the scalefactors are incorrect, you will end up with misaligned coordinates. This could for example happen if you are trying to use the orignal H&E image and not the "tissue_hires_image.png" when running `InputFromTable`.

In this vignette we have used the datasets from Mouse Brain Serial Section 1 and 2 (Sagittal-Posterior)

### Prepare data

The recommended method to read the files into R is via the creation of a `data.frame` that we will call the `infoTable`. There are four columns in this table that are required for Visium data: "samples", "spotfiles", "imgs" and "json". These columns specify the paths to the required input files.

<br>
```{r example_infotable_Visium, echo=FALSE, eval=TRUE}

samples <- c("path/to/sample_1/count_file_1.h5 ", "path/to/sample_2/count_file_2.h5")
spotfiles <- c("path/to/sample_1/tissue_positions_list.csv", "path/to/sample_2/tissue_positions_list.csv")
imgs <- c("path/to/sample_1/tissue_hires_image.png", "path/to/sample_2/tissue_hires_image.png")
json <- c("path/to/sample_1/scalefactors_json.json", "path/to/sample_2/scalefactors_json.json")
infoTable <- data.frame(samples, spotfiles, imgs, json)

infoTable %>%
  kbl() %>%
  kable_paper() %>%
  scroll_box(width = "800px")
```

Any number of _extra_ columns can be added to the infoTable data.frame that you want to include as meta data in your Seurat object, e.g. "gender", "age", "slide_id" etc. These columns can be named as you like, but they should not be called _not_ "sample", "spotfiles", "imgs" or "json".

We are now ready to load our samples and create a "Seurat" object using our infotTable `data.frame`. 

Here, we demonstrate the creation of the Seurat object, while also including some filtering (see section "Quality Control" for more information on filtering): 

* Keeping the genes that are found in at least 5 capture spots and has a total count value >= 100. 
* Keeping the capture-spots that contains >= 500 total transcripts. 

Note that you have to specify which platform the data comes from. The default platform is 10X Visium but if you wish to run data from the older ST platforms, there is support for "1k" and "2k" arrays. You can also mix datasets from different platforms by specifying one of; "Visium", "1k" or "2k" in a separate column of the infoTable named "platform". You just have to make sure that the datasets have gene symbols which follows the same nomenclature so that count matrices can be merged.
<br>
```{r package_infotable_Visium, eval=FALSE, include = FALSE}

samples <- c("~/10x/Public_data/S1_Sagittal_posterior/V1_Mouse_Brain_Sagittal_Posterior_filtered_feature_bc_matrix.h5",
             "~/10x/Public_data/S2_Sagittal_posterior/V1_Mouse_Brain_Sagittal_Posterior_Section_2_filtered_feature_bc_matrix.h5",
             "~/10x/Public_data/S1_Sagittal_anterior/V1_Mouse_Brain_Sagittal_Anterior_filtered_feature_bc_matrix.h5",
             "~/10x/Public_data/S2_Sagittal_anterior/V1_Mouse_Brain_Sagittal_Anterior_Section_2_filtered_feature_bc_matrix.h5")
imgs <- c("~/10x/Public_data/S1_Sagittal_posterior/tissue_hires_image.png",
             "~/10x/Public_data/S2_Sagittal_posterior/tissue_hires_image.png",
             "~/10x/Public_data/S1_Sagittal_anterior/tissue_hires_image.png",
             "~/10x/Public_data/S2_Sagittal_anterior/tissue_hires_image.png")
spotfiles <- c("~/10x/Public_data/S1_Sagittal_posterior/tissue_positions_list.csv",
             "~/10x/Public_data/S2_Sagittal_posterior/tissue_positions_list.csv",
             "~/10x/Public_data/S1_Sagittal_anterior/tissue_positions_list.csv",
             "~/10x/Public_data/S2_Sagittal_anterior/tissue_positions_list.csv")
json <- c("~/10x/Public_data/S1_Sagittal_posterior/scalefactors_json.json",
             "~/10x/Public_data/S2_Sagittal_posterior/scalefactors_json.json",
             "~/10x/Public_data/S1_Sagittal_anterior/scalefactors_json.json",
             "~/10x/Public_data/S2_Sagittal_anterior/scalefactors_json.json")
infoTable <- data.frame(samples, imgs, spotfiles, json, stringsAsFactors = FALSE)
saveRDS(object = infoTable, file = "../pre_data/infoTable_filtered")

```
```{r read_infotable, include=FALSE}

infoTable <- readRDS(file = "pre_data/infoTable_filtered")

```
```{r input_from_table_visium, eval=F}

se <- InputFromTable(infotable = infoTable, 
                      min.gene.count = 100, 
                      min.gene.spots = 5,
                      min.spot.count = 500,
                      platform =  "Visium")
```

```{r load_image_Visium, include=FALSE, eval=F}

se <- LoadImages(se, time.resolve = FALSE)

```

```{r pre-load, include=FALSE, eval=TRUE}

load("pre_data/preSaved_10x_serial.RData")

```
<br>
Once you have created a Seurat object you can process and visualize your data just like in a scRNA-seq experiment and make use of the plethora of functions provided in the `Seurat` package. There are many vignettes to get started available at the [Seurat web site](https://satijalab.org/seurat/vignettes.html).

For example, if you wish to explore the spatial distribution of various features on the array coordinates you can do this using the `ST.FeaturePlot()` function. Features include any column stored in the "meta.data" slot, dimensionality reduction objects or gene expression vectors. 

<br>


```{r, eval=TRUE, fig.width=10, fig.height=10, out.width='100%'}

ST.FeaturePlot(se, features = c("nFeature_RNA"), cols = c("lightgray", "mistyrose", "red", "darkred", "black"), ncol = 2, pt.size = 1.3)

```

<br>

## Older ST platforms

In general, using STutility for the old ST platform data follows the same workflow as for the 10X Visium arrays. The only difference is when loading the data into R.

### Input files

The original ST workflow produces the following three output files:

1. Count file in TSV format (Count file with raw counts (UMI filtered) for each gene and capture spot)
2. Spot detector output (File with spatial pixel coordinate information produced using the [Spot Detector webtool](https://github.com/SpatialTranscriptomicsResearch/st_spot_detector))
3. H&E image (same as the one used for spot detection)

### Prepare data

The recommended method to read the files into R is via the creation of a "infoTable", which is a table with at least three columns "samples", "spotfiles" and "imgs". 

Test data:

```{r package_infotable, eval=FALSE}

samples <- list.files(pattern = "^Hippo[1-6].tsv.gz", path = system.file("extdata/counts", package = "STutility"), full.names = TRUE)
spotfiles <- list.files(pattern = "^alignment_table_Hippo[1-6].tsv.gz", path = system.file("extdata/counts", package = "STutility"), full.names = TRUE)
imgs <- list.files(pattern = "^Hippo[1-6].jpg", path = system.file("extdata/counts", package = "STutility"), full.names = TRUE)

infoTable <- data.frame(samples, imgs, spotfiles, stringsAsFactors = F)

```

### Load data and convert from EnsambleIDs to gene symbols

The provided count matrices uses EnsambleIDs (with version id) for the gene symbols. Gene symbols are often a preference for easier reading, and we have therefore included an option to directly convert the gene IDs when creating the Seurat object. The data.frame object required for conversion should have one column called "gene_id" matching the original gene IDs and a column called "gene_name" with the desired symbols. you also need to make sure that these columns have unique names, otherwise the converiion will not work. We have provided such a table that you can use to convert between EnsambleIDs and MGI symbols (mus musculus gene nomenclature). 

<br>
```{r annotation, eval=FALSE}

#Transformation table for geneIDs
ensids <- read.table(file = list.files(system.file("extdata", package = "STutility"), full.names = T, pattern = "mouse_genes"), header = T, sep = "\t", stringsAsFactors = F)

```

We are now ready to load our samples and create a "Seurat" object. 

Here, we demonstrate the creation of the Seurat object, while also including some filtering: 

* Keeping the genes that are found in at least 5 capture spots and has a total count value >= 100. 
* Keeping the capture-spots that contains >= 500 total transcripts. 

Note that we specify that we're using the "2k" array platform and also, since we in this case have genes in the columns, we set transpose=TRUE. 
<br>
```{r read_input_1, eval=FALSE}

#TODO: add warnings if ids missmatch. Check that ids are in the data.frame ...
se <- InputFromTable(infotable = infoTable, 
                      min.gene.count = 100, 
                      min.gene.spots = 5,
                      min.spot.count = 500, 
                      annotation = ensids, 
                      platform = "2k",
                      transpose = TRUE)

```
<br>
Once you have created a Seurat object you can process and visualize your data just like in a scRNA-seq experiment and make use of the plethora of functions provided in the Seurat package. There are many vignettes to get started available at the [Seurat web site](https://satijalab.org/seurat/vignettes.html).

Some of the functionalities provided in the Seurat package are not yet supported by STutility, such as dataset integration and multimodal analysis. These methods should in principle work if you treat the data like a scRNA-seq experiment, but you will not be able to make use of the image related data or the spatial visualization functions.

For example, if you wish to explore the spatial distribution of various features on the array coordinates you can do this using the `ST.FeaturePlot()` function.
<br>
```{r load_image, include=FALSE, eval=FALSE}

se <- LoadImages(se, time.resolve=FALSE)

```

```{r pre-load_2k, include=FALSE, eval=TRUE}

load("pre_data/preSaved_2kHippo.RData")

```

```{r pot_fts, fig.width=8, fig.height=8, out.width='100%'}

ST.FeaturePlot(se, features = c("nFeature_RNA"), cols = c("lightgray", "mistyrose", "red", "darkred", "black"), ncol = 2, pt.size = 1.5, show.sb = FALSE)

```
<br>

## Navigating and accessing data
***

We recommend that unexperienced users have look at the `Seurat` website and tutorials for basic navigation of the Seurat object such as getting and setting identities and accessing various method outputs.

However, specific for STutility, there is another S4 class object stored inside the `Seurat` object, called `Staffli`. This object contains all the STutility specific meta data, like pixel cooridinates, sample IDs, platform types etc.

You can reach this via:

```{r staffli}

st.object <- GetStaffli(se)
st.object

# Access meta data
head(st.object[[]]) %>%
  kbl() %>%
  kable_styling()

```

It can for example be useful to access the spot coordinates and images if you want to write your own plotting functions.

### Important note
***

Because the spatial infromation is stored in a slot that is normally not part of the `Seurat` environment, some function such as `subset` or `merge` will not work as expected. Instead, you should use the `SubsetSTData` and `MergeSTData` functions from STutility.

&nbsp;
<hr />
<p style="text-align: center;">A work by <a href="mailto: j.bergenstrahle@scilifelab.se">Joseph Bergenstråhle</a> and <a href="mailto: ludvig.larsson@scilifelab.se">Ludvig Larsson</a></p>

<!-- Add icon library -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">

<!-- Add font awesome icons -->
<p style="text-align: center;">
    <a href="https://www.spatialresearch.org" class="fa fa-beer"></a>
</p>

&nbsp;
