---
title: "STUtility - Vignette"
author:
- Joseph Bergenstråhle, Royal Institute of Technology (KTH)
- Ludvig Larsson, Royal Institute of Technology (KTH)
date: ''
#output:
#  html_document:
#    theme: darkly
#    css: style.css
#    number_sections: false
#    toc: yes
#    toc_depth: 3
#    toc_float:
#      collapsed: false
#      smooth_scroll: true
#  pdf_document:
#    number_sections: false
#    toc: yes
#    toc_depth: 3
---

<style type="text/css">
div.main-container {
  background-color: #000000 !important;
  max-width: 1400px;
  margin-left: auto;
  margin-right: auto;
}
</style>
<style>
#TOC {
  background: url("https://www.spatialresearch.org/wp-content/uploads/2019/09/str-logo-spatial_research_3@2x.png");
  background-size: contain;
  padding-top: 100px !important;
  background-repeat: no-repeat;
  op: 5%;
  opacity: 0.8;
  width: 500px;
  color: white;
  border-color: #000000 !important;
}
</style>

<style> code, pre{
  background-color: #000000 !important;
  color: white !important;
}
</style>
<style> 
body {
  color: white
}
</style>
<style>
.list-group-item.active, .list-group-item.active:focus, .list-group-item.active:hover {
    background-color: #375a7f;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, autodep = TRUE)
```

```{r load_lib, warning=FALSE, message=FALSE, results="hide", autodep=TRUE, include=FALSE}
library(STutility)
```

## 10X Visium platform

### Input files

After running the 10X SpaceRanger pipeline, the output includes:

1. Count file in H5 format
2. Barcode position file 
3. H&E image

To use the full range of functions within STutility, all three files are needed for each sample. However, all data analysis steps that do not involve the H&E image can be performed with only the count file as input. 

---

To follow along this tutorial, download the test data set at TODO:[insert test data set link]. The downloadable content consists of count files, output from our spot detector tool, H&E stained images as well as an "infoTable" to read in the files into R.

### 10x infoTable

The same format is used as the original ST platform, see \@ref(The-infoTable)
<br>
```{r infotable2, cache=TRUE}

samples <- list.files(path = "~/Hippocampus_Visium/", recursive = T, pattern = "filtered_feature_bc_matrix.h5", full.names = T)
imgs <- list.files(path = "~/Hippocampus_Visium/", recursive = T, pattern = "hires_image.png", full.names = T)
spotfiles <- list.files(path = "~/Hippocampus_Visium/", recursive = T, pattern = "list.txt", full.names = T)

infoTable <- data.frame(imgs, samples, spotfiles, sample_id = c("Hippo_A", "Hippo_B"), stringsAsFactors = F)

```

## Loading 10X data

To read in the 10x Visium .h5 files, the package `hdf5r` needs to be installed (`BiocManager::install("hdf5r")`). By default 10X Visium output format implies setting `transpose=F`. [There is a specified scaling factor for the H&E image. The picture with suffix "high_res" uses the default scaling factor of 0.1039393. However, is other pictures are used, specify the scaling factor when calling the _InputFromTable()_ function. All scaling factors are found in the scalefactors_json.json which you have in your 10X output folder] (OBS CHECK HOW THE WORKFLOW 10X IS SUPPOSED TO BE)
<br>
```{r read_input_2, cache=TRUE}

se <- InputFromTable(infotable = infoTable, 
                      min.gene.count = 100, 
                      min.gene.spots = 5,
                      min.spot.count = 500)

```


# Loading and handling images

With the Seurat object created, we are ready to load the H&E images from the file paths provided in the infoTable. The LoadImages() function allow you to load the images into the Seurat object and will automatically save a scaled down version of each image that you can use for plotting.

You should be aware that the resolution is by default very low (400 pixels wide). The reason for this is that STUtility makes it possible to apply various image transformations to the images, but since these processing steps are slow and requires significant amounts of memory, it is much faster to learn the transformations on a small image and later apply the transformations to a higher resolution image. See the image processing section of this vignette for more details.

<br>
```{r load_images, echo=TRUE, eval=FALSE}

se <- LoadImages(se, time.resolve = F, verbose = T)

```
<br>
OBS: We have noted that some systems experience memory problems due to large size image information not being cleared when removing old objects. If this is the case, the user can set `time.resolve=TRUE`, which will add a small garbage collection step to avoid this issue.

NOTE: time.resolve is set to TRUE as default in the current version. This will imply that the process takes approx. 5sec longer per sample. Turn this off if annoyed. 

---

The core of an ST experiment is the interplay between quantitative RNAseq data and qualitative image data. Working in R, the handling of image information and plotting of multiple samples can get a bit cumbersome, especially if the images are of higher resolution. Typically, we rarely find much use of plotting multiple samples together with the tissue images within this framework. Instead, the general tip is to perform the various plots using only the expression data, and after finding anything of interest that warrants further analysis - take out that sample and plot it individually together with the stained tissue image. 

For example, we can visualize the number of RNA features and RNA counts on the ST array coordinates and quickly verify where we capture more material from the tissue.

NOTE: see the end of the vignette how to handle plotting if you are missing image/spot-files. 
<br>
```{r load_se, include=FALSE}

# Read in the processed Seurat object so that all computationally demanding downstream steps can be skipped
se <- readRDS(file = "~/STUtility/se_object")

```

```{r plot_features1, fig.height=12, fig.width=12, out.width = '100%'}
# TODO: make sure that levels are set in facets

ST.FeaturePlot(object = se, 
               features = c("nFeature_RNA", "nCount_RNA"), 
               cols = c("black", "dark blue", "cyan", "yellow", "red", "dark red"),
               ncol = 2, 
               grid.ncol = 1, 
               pt.size = 0.5, 
               dark.theme = T)

```
<br>
To overlay a spatial feature heatmaps on top of the HE image we use the FeatureOverlay() function. Handling large images gets clumsy so we recommend to work with down-scaled versions of the images that takes up less memory.
<br>

```{r overlay_features1, fig.width = 6, fig.height = 6, out.width = '100%'}

FeatureOverlay(se, 
               features = "Cck", 
               pt.size = 0.5, 
               pt.alpha = 0.5, 
               cols = c("black", "dark blue", "cyan", "yellow", "red", "dark red"), 
               dark.theme = T, 
               type = "raw")

```
<br>
You can see that by default, the first sample was selected for the feature overlay. If you wish to visualize another sample you can specify the `sample.index` parameter. If you wish to plot several features side by side, you can just specify as many as you want.
<br>
```{r overlay_features_2, fig.width = 12, fig.height = 12, out.width = '100%'}

FeatureOverlay(se, 
               features = c("Cck", "Calb2", "Dcn", "Nrgn"), 
               pt.size = 0.5,
               pt.alpha = 0.5, 
               cols = c("black", "dark blue", "cyan", "yellow", "red", "dark red"), 
               dark.theme = T, 
               type = "raw")

```

<br>
Another neat feature that can be useful for visualization purposes is to mask the background of the HE images. The function `MaskImages()` can be used to remove the background currently works well for tissues with a clearly defined border. Since masking is a non trivial problem, it is possible that the masking will fail for certain tissue types. If this happens you can try to modify the parameters in `MaskImages()` to see if you can obtain better results. 

NOTE that the image procesing methods have been optimized for small images and if you have loaded images in higher resolution there is a greater risk that some steps will fail. If you still wish to use images with higher resolution there are instructions on how to do this in the Resolution section below.
<br>
```{r mask_images, eval=FALSE}

se <- MaskImages(object = se)

```
<br>
There are 5 types of images which can be stored into the Seurat object; "raw", "masked", "masked.masks", "processed" and "processed.masks". "raw" represents the original HE images without any modifications. In the "masked" images, the background have been removed using a mask which is stored in "masked.masks". Similarly, the "processed"  images are also masked using "processed.masks" but have been further modified by applying rigid transformations such as rotations and shifts.

You can look at the HE images using the ImagePlot() function.The ImagePlot function will automatically label the samples from 1 to N where N is the total number of samples. If you are running the function within RStudio you can take advantage of the Viewer panel to explore the images. Setting method="raster" and the plot function will use the default plotting device in R. You can also change the layout of the images by forcing a different number of columns by setting ncols.
<br>
```{r image_types, fig.width=11, fig.height=6, out.width = '100%'}

# Plot HE images
ImagePlot(se, ncols = 2, method = "raster", type = "raw", darken = T) # Raw image
ImagePlot(se, ncols = 2, method = "raster", type = "masked", darken = T) # Masked image
ImagePlot(se, ncols = 2, method = "raster", type = "masked.masks") # Mask

```

### Individual transformations to images

If you follow along the test set in this vignette, you will notice that the samples are not quite aligned with respect to each other on the array. We have implemented three different method to align images; `AlignImages()`, `WarpImages()` and `ManualAlignImages()`.

#### Automatic alignment

The automatic alignment method (`AlignImages()`) first tries to detect the tissue edges from each image. By default, the first image (`reference.index = 1`) is used as a reference but you can specify any sample with the `reference.index`option. Then, for each image, a transformation matrix is learned that can be used to map coordinates to the reference image. This alignment method will sometimes fail and in that case you can align the images manually instead using the `ManualAlignImages()` function. This is typically required when; (1) the tissue is bigger than the image and therefore outside the frame, (2) the tissue has a symmetrical shape (for example if the tissue has a round shape, it will be difficult to find an optimal alignment using the tissue edges), (3) the masking has failed.

The actual transformation is done using the `imwarp()` function from the imager R package using a "backward" transformation strategy. This method makes sure that every pixel is painted using linear interpolation so the aligned images will experience some quality loss.
<br>
```{r align_images, fig.width=12, fig.height=12, eval = FALSE}

se <- AlignImages(se)

```
<br>
And the result:
<br>
```{r imageplot_1, fig.width=11, fig.height=6, out.width = '100%'}

ImagePlot(se, method = "raster", darken = T)

```


#### Manual rigid transformations

An alternative method is to apply rigid transformations to the images using the `WarpImages()` function. This function takes a list of transformations for the samples of your choice and converts the transformations into mapping functions which can be used to warp the images. 

The available transformations are; "center.x", "center.y", "mirror.x", "mirror.y" and "angle"
<br>
```{r warpimages, fig.width=12, fig.height=6, out.width = '100%'}

transforms <- list("2" = list("angle" = 90))
se.rotate90 <- WarpImages(se, transforms)
ImagePlot(se.rotate90, method = "raster", darken = T)

transforms <- list("2" = list("mirror.x" = T))
se.mirrorx <- WarpImages(se, transforms)
ImagePlot(se.mirrorx, method = "raster", darken = T)

transforms <- list("2" = list("mirror.y" = T))
se.mirrory <- WarpImages(se, transforms)
ImagePlot(se.mirrory, method = "raster", darken = T)

```
<br>
Any transformations applied to the images will also be applied to the coordinates used to map gene expression values onto the images. 
<br>
```{r show_warp, fig.height=6, fig.width=12, out.width = '100%'}

heatmap.colors <- c("black", "dark blue", "cyan", "yellow", "red", "dark red")
ST.FeaturePlot(se, features = "Cck", pt.size = 0.5, cols = heatmap.colors, dark.theme = T)
ST.FeaturePlot(se.rotate90, features = "Cck", pt.size = 0.5, cols = heatmap.colors, dark.theme = T)
ST.FeaturePlot(se.mirrorx, features = "Cck", pt.size = 0.5, cols = heatmap.colors, dark.theme = T)
ST.FeaturePlot(se.mirrory, features = "Cck", pt.size = 0.5, cols = heatmap.colors, dark.theme = T)

```

### Visualize multiple samples

If you still want to plot overlaid spatial heatmaps for multiple samples at the same time you can do this with the `MultiFeatureOverlay()` function. 
<br>
```{r overlay_multi, fig.width=12, fig.height = 6, out.width = '100%'}

MultiFeatureOverlay(se, features = "Nrgn", 
                    sampleids = 1:2,
                    pt.size = 0.5, pt.alpha = 0.5,
                    cols = c("black", "dark blue", "cyan", "yellow", "red", "dark red"),
                    method = "raster",
                    dark.theme = T)

```
<br>
The `type` parameter can also be set in the `FeatureOverlay()` function to override the default use of processed images. 
<br>
```{r overlay_multi_raw, fig.height = 6, fig.width = 12, out.width = '100%'}

FeatureOverlay(object = se, features = c("Nrgn", "Cck"), 
               cols = c("black", "dark blue", "cyan", "yellow", "red", "dark red"), 
               pt.size = 0.5, 
               pt.alpha = 0.5,
               sample.index = 2,
               type = "raw", 
               dark.theme = T)

```

&nbsp;
<hr />
<p style="text-align: center;">A work by <a href="j.bergenstrahle@scilifelab.se">Joseph Bergenstråhle</a> and <a href="ludvig.larsson@scilifelab.se">Ludvig Larsson</a></p>

<!-- Add icon library -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">

<!-- Add font awesome icons -->
<p style="text-align: center;">
    <a href="https://www.spatialresearch.org" class="fa fa-beer"></a>
</p>

&nbsp;
