---
title: "Spatial Features"
author:
date: ''
#output:
#  html_document:
#    theme: darkly
#    css: style.css
#    number_sections: false
#    toc: yes
#    toc_depth: 3
#    toc_float:
#      collapsed: false
#      smooth_scroll: true
#  pdf_document:
#    number_sections: false
#    toc: yes
#    toc_depth: 3
---

<style type="text/css">
div.main-container {
  background-color: #000000 !important;
  max-width: 1400px;
  margin-left: auto;
  margin-right: auto;
}
</style>
<style>
#TOC {
  background: url("https://www.spatialresearch.org/wp-content/uploads/2019/09/str-logo-spatial_research_3@2x.png");
  background-size: contain;
  padding-top: 100px !important;
  background-repeat: no-repeat;
  op: 5%;
  opacity: 0.8;
  width: 500px;
  color: white;
  border-color: #000000 !important;
}
</style>

<style> code, pre{
  background-color: #000000 !important;
  color: white !important;
}
</style>
<style> 
body {
  color: white
}
</style>
<style>
.list-group-item.active, .list-group-item.active:focus, .list-group-item.active:hover {
    background-color: #375a7f;
}
</style>



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, autodep = TRUE)

```

```{r load_lib, warning=FALSE, message=FALSE, results="hide", autodep=TRUE, include=FALSE}
library(STutility)
#load("~/STUtility/saved/preSaved_10xHippo_norm_reductions.RData")

se.all <- readRDS("F:/JoeDocs/PhD/STUtility_website/se.list")
se <- se.all[[1]]
```

## Finding spatial expression patterns

The strength of untargeted whole transcriptome capture is the ability to perform unsupervised analysis and the ability to find spatial gene expression patterns. We've found good use of using non-negative matrix factorization (NMF) to find underlying patterns of transcriptomic profiles. This factor analysis, along with various dimensionality reduction techniques, can all be ran via "RunXXX()", where X = the method of choice, e.g.:
<br>
```{r NMF, fig.width=16, fig.height=10, eval = FALSE}

se <- RunNMF(se, nfactors = 40) #Specificy nfactors to choose the number of factors, default=20.

``` 
<br>
While `RunNMF()` is an STUtility add-on, others are supported via Seurat (`RunPCA()`, `RunTSNE`, `RunICA()`, `runUMAP()` ) and for all of them, the output are stored in the Seurat object. 

We can then plot a variable number of dimensions across the samples.
<br>
```{r dimplot, fig.height=28, fig.width=12, out.width = "100%"}

cscale <- c("black", "darkblue", "cyan", "yellow", "red", "darkred")

ST.DimPlot(se, 
           dims = 1:10,
           ncol = 2, # Sets the number of columns at dimensions level
           grid.ncol = 2, # Sets the number of columns at sample level
           reduction = "NMF", 
           dark.theme = T, 
           pt.size = 0.5, 
           center.zero = F, 
           cols = cscale)

ST.DimPlot(se, 
           dims = 11:20,
           ncol = 2, 
           grid.ncol = 2, 
           reduction = "NMF", 
           dark.theme = T, 
           pt.size = 0.5, 
           center.zero = F, 
           cols = cscale)


ST.DimPlot(se, 
           dims = 21:30,
           ncol = 2, 
           grid.ncol = 2, 
           reduction = "NMF", 
           dark.theme = T, 
           pt.size = 0.5, 
               center.zero = F, 
           cols = cscale)

ST.DimPlot(se, 
           dims = 31:40,
           ncol = 2, 
           grid.ncol = 2, 
           reduction = "NMF", 
           dark.theme = T, 
           pt.size = 0.5, 
               center.zero = F, 
           cols = cscale)

```
<br>
To extract the genes that drives the separation according to the dimensionality reduction, we can use the `ProjectDim()` function.
<br>
```{r project_dim}

ProjectDim(se, reduction = "NMF", dims = c(2,4))

```


## Clustering

Clustering is a standard procedure in genomic analysis, and the methods for doing so are numerous. Here we demonstrate an example where we use the result of the factor analysis the previous section. Going through the list of factors (e.g. via `ST:DimPlot(se, dims = [dims you want to look at])`), we can notice dimensions that are "spatially active", i.e. that seems to confer a spatial pattern along their axis. We can extract these dimensions and use as input to e.g. clustering functions. Here, we use all dimensions from the NMF and construct a Shared Nearest Neighbor (SSN) Graph. 
<br>
```{r findneighbours, eval=FALSE}
se <- FindNeighbors(object = se, verbose = FALSE, reduction = "NMF", dims = 1:40)
```
<br>
Followed by clustering using a modularity optimizer
<br>
```{r findclusters, eval=FALSE}
se <- FindClusters(object = se, verbose = FALSE)
```
<br>
And plotting of the clusters spatially
<br>
```{r plot_clusters, fig.height = 10, fig.width = 12, out.width='100%'}

library(RColorBrewer)
n <- 19
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))

ST.FeaturePlot(object = se, features = "seurat_clusters", dark.theme = T, cols = col_vector, pt.size = 0.5)

```
<br>
If you think that the distribution of clusters gets too cluttered, you can also split the view so that only one cluster at the time gets colored.
<br>
```{r plot_clusters_split, fig.height=18, fig.width=22, out.width="100%"}

ST.FeaturePlot(object = se, features = "seurat_clusters", dark.theme = T, pt.size = 1, split.labels = T, indices = 1)


ST.FeaturePlot(object = se, features = "seurat_clusters", dark.theme = T, pt.size = 1, split.labels = T, indices = 3)
```
<br>

## Most variable features

We can take a specific look at some of the most variable features.
<br>
```{r variable_features, fig.height=12, fig.width=6, out.width='100%'}

head(se@assays$SCT@var.features, 20)
top <- se@assays$SCT@var.features

fts <- c("Ttr", "Mbp", "Nrgn")
for (ftr in fts) {
  FeatureOverlay(se, 
                    features = ftr, 
                    sampleids = 1:2,
                    cols = c("black", "darkblue", "cyan", "yellow", "red", "darkred"),
                    pt.size = 0.5, 
                    pt.alpha = 0.5, 
                    dark.theme = T)
}

```
<br>

## Compare graph embeddings

Another useful feature is that you can now compare the spatial distribution of a gene with the typical "graph embeddings" s.a. UMAP and t-SNE. 
<br>
```{r run_UMAP, eval=F}

# Run UMAP
se <- RunUMAP(se, reduction = "NMF", n.neighbors = 10)

````

```{r embedding_vs_ST, fig.width = 9, fig.height = 14}

# Define colors for heatmap
heatmap.colors <- c(rgb(40, 40, 40, maxColorValue = 255), "darkblue", "cyan", "white")
fts <- c("Prkcd", "Opalin", "Lamp5")

# plot transformed features expression on UMAP embedding
p.fts <- lapply(fts, function(ftr) {
  FeaturePlot(se, features = ftr, reduction = "umap", order = TRUE, cols = heatmap.colors) + DarkTheme()
})

# plot transformed features expression on Visium coordinates
p3 <- ST.FeaturePlot(se, features = fts, ncol = 2, grid.ncol = 1, cols = heatmap.colors, pt.size = 0.5, dark.theme = T)

# Construct final plot
cowplot::plot_grid(cowplot::plot_grid(plotlist = p.fts, ncol = 1), p3, ncol = 2, rel_widths = c(1, 1))

```

## RGB dimensionality reduction plots

One approach to visualize the result of dimensionality reduction is to use the first three dimensions and transform the values into RGB color space. This 3 dimensional space can then be utilized for spatial visualization.
We demonstrate the technique with UMAP, using our factors as input:
<br>
```{r UMAP, val=FALSE, eval=T, warning=F}

se <- RunUMAP(object = se, dims = 1:40, verbose = FALSE, n.components = 3, reduction = "NMF", reduction.name = "umap.3d")

```
<br>
We use the first three dimensions for plotting:
<br>
```{r UMAP_blend, fig.height = 8, fig.width = 10}

ST.DimPlot(object = se, dims = 1:3, reduction = "umap.3d", blend = T, dark.theme = T, pt.size = 0.5)

```

<br>
Once again the `ProjectDim()` function can be used to display the genes that are most strongly correlated with the coordinate system. Note in the function call above that we defined `reduction.name`, which are subsequently stored in the Seurat object in the reduction slot: 
<br>
```{r projectdim_UMAP}

ProjectDim(se, reduction = "umap.3d")

```
<br>

## DEA

Lets try this out by an example. Looking at \ref{plot_clusters_split}, lets say we are interested in cluster 19, and want to answer - "Which genes are significantly up-regulated in this region compared to the overall expression across the tissue?"

```{r}

markers <- FindMarkers(se, ident.1 = "19")
head(markers)
```

Note that the clusters were already set as the Seurat objects levels. Type `levels(se)` to see the current levels of your object. If other clusters, annotations etc are of interest, set this before by specifying `Idents(se) <- `

Note also, if we are interested in comparing two levels against each other, and not just "one against the rest", we simply add a `ident.2 = ` parameter to the above. 


```{r, fig.height=12, fig.width=6}

FeatureOverlay(se, features = "Dlx1", 
                    sampleids = 1:2,
                    cols = c("black", "darkblue", "cyan", "yellow", "red", "darkred"),
                    pt.size = 1, 
                    pt.alpha = 0.5,
                    dark.theme = T)

```
<br>

## Spatial Auto-correlation

## Spatial Auto-correlation

STUtility also includes an additional method for finding genes with spatial patterns across the tissue. The ranking method makes use neighborhood networks to compute the spatial lag for each gene, here defined as the summed expression of that gene in neighboring spots. Each gene is then ranked by the correlation between the lag vector and the original expression vector. The output is a ranked list of the genes with highest correlation between the two vectors. 

This method is inspired by work from the [Giotto team](http://spatial.rc.fas.harvard.edu/spatialgiotto/giotto.html) and we reccomend you to check out their R package Giotto and their preprint; ["Giotto, a pipeline for integrative analysis and visualization of single-cell spatial transcriptomic data"](https://www.biorxiv.org/content/10.1101/701680v1) that is available on bioRxiv.

```{r autocorrelation, eval=FALSE}

CorSpatialGenes(se)

```


## Region neighbors

```{r, eval=T, echo=F}

library(magrittr)
library(dplyr)

se <- se.all[[4]]
```

To demonstrate the regional neighbors feature, we will use the Breast Cancer data available on the 10x Genomics [website](https://support.10xgenomics.com/spatial-gene-expression/datasets).

### Connected Spatial Network

If you have defined a region of interest and you want to find all spots neighboring to this region you can use the RegionNeighbours() function to automatically detect such spots. For example, let’s say that we want to select all neighbors to cluster 12 in the plot below and find the differential expressed markers between the neighbors.

```{r, echo=T,  fig.width = 14, fig.height=7, out.width = "100%"}
cols <- c("#771155", "#AA4488", "#CC99BB", "#114477", "#4477AA", "#77AADD", "#117777", "#44AAAA", "#77CCCC", "#117744", "#44AA77", "#88CCAA", "#777711", "#AAAA44", "#DDDD77", "#774411", "#AA7744", "#DDAA77")

ST.FeaturePlot(se, features = "seurat_clusters", pt.size = 1, cols=cols, dark.theme = T)


```

First, RegionNeighbours() function to find the neighboring spots but we have to make sure that the active identity is appropriately set, here we use ‘seurat_clusters’.

```{r}

se <- SetIdent(se, value = "seurat_clusters")
se <- RegionNeighbours(se, id = 2, verbose = TRUE)

```

The default behavior is to find all spots which are neighbors with the selected id but ignoring its label, therefore it will simply be called nbs_12 as in “neighbors to 12”.

```{r, echo=T, fig.width = 14, fig.height=6, out.width = "100%"}

FeatureOverlay(se, features = "nbs_2", ncols.samples = 2, sampleids = 1:2, cols = c("red", "black", "lightgray"), pt.size = 1, dark.theme = T)

```

You can also keep all spots from the id group by setting keep.within.id = TRUE.

```{r}

se <- SetIdent(se, value = "seurat_clusters")
se <- RegionNeighbours(se, id = 2, keep.within.id = T, verbose = TRUE)

```

```{r, echo=T, fig.width = 14, fig.height=6, out.width = "100%"}

FeatureOverlay(se, features = "nbs_2", ncols.samples = 2, sampleids = 1:2, cols = c("red", "black", "lightgray"), pt.size = 1, dark.theme = T)

```

```{r}

se <- SetIdent(se, value = "nbs_2")
nbs_2.markers <- FindMarkers(se, ident.1 = "2", ident.2 = "nbs_2")
nbs_2.markers$gene <- rownames(nbs_2.markers)
se.subset <- SubsetSTData(se, spots = rownames(subset(se[[]], nbs_2 %in% c("2", "nbs_2"))))
sorted.marks <- nbs_2.markers %>% top_n(n = 40, wt = abs(avg_logFC))
sorted.marks <- sorted.marks[order(sorted.marks$avg_logFC, decreasing = T), ]
DoHeatmap(se.subset, features = sorted.marks$gene, group.colors = c("black", "lightgray"), disp.min = -2, disp.max = 2)+ DarkTheme() 

```

And lastly, if you want to keep the labels for the neighbors you can set keep.idents = TRUE and the neighbors will be called label_nb_to_12.

```{r, echo=T, fig.width = 14, fig.height=6, out.width = "100%"}

se <- SetIdent(se, value = "seurat_clusters")
se <- RegionNeighbours(se, id = 2, keep.idents = TRUE, verbose = TRUE)
FeatureOverlay(se, features = "nbs_2", ncols.samples = 2, sampleids = 1:2, pt.size = 1, dark.theme = T)

```

<br>

&nbsp;
<hr />
<p style="text-align: center;">A work by <a href="j.bergenstrahle@scilifelab.se">Joseph Bergenstråhle</a> and <a href="ludvig.larsson@scilifelab.se">Ludvig Larsson</a></p>

<!-- Add icon library -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">

<!-- Add font awesome icons -->
<p style="text-align: center;">
    <a href="https://www.spatialresearch.org" class="fa fa-beer"></a>
</p>

&nbsp;
