---
title: "Spatial Features"
author:
date: ''
---

<style type="text/css">
div.main-container {
  background-color: #FFFFFF !important;
  max-width: 1800px;
  margin-left: auto;
  margin-right: auto;
}
</style>

<style>
#TOC {
  background: url("https://www.spatialresearch.org/wp-content/uploads/2019/09/str-logo-spatial_research_3@2x.png");
  background-size: contain;
  padding-top: 100px !important;
  background-repeat: no-repeat;
  width: 500px;
  color: black;
  border-color: #FFFFFF !important;
}
</style>

<style> code, pre{
  background-color: #F8F8F8 !important;
  color: black !important;
}
</style>

<style> .watch-out {
  background-color: #F0F0F0;
  border: 1px #F0F0F0;
}
<style>

<style> 
body {
  color: black;
  font-size: 20pt;
}
</style>


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, autodep = TRUE)

```

```{r load_lib, warning=FALSE, message=FALSE, include=FALSE}
library(STutility)
```
```{r load_lib_hidden, include=FALSE}
library(kableExtra)
library(magrittr)
```

```{r, pre-load, echo=FALSE, eval=TRUE, include=FALSE}

load("pre_data/preSaved_10xHippo_norm_reductions.RData")

```

## Spatial Auto-correlation
***

STutility includes a method for finding genes with spatially conserved patterns across the tissue. The ranking method makes use neighborhood networks to compute the spatial lag for each gene, here defined as the summed expression of that gene in neighboring spots. Each gene is then ranked by the correlation between the lag vector and the original expression vector. The output is a `data.frame` with gene names ranked by decreasing spatial auto-correlation. 

This method is partly inspired by work from the [Giotto team](http://spatial.rc.fas.harvard.edu/spatialgiotto/giotto.html) and we reccomend you to check out their R package Giotto and the related publication; ["Giotto: a toolbox for integrative analysis and visualization of spatial expression data"](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02286-2.

<br>
```{r autocorrelation, eval=FALSE}

library(spdep)
spatgenes <- CorSpatialGenes(se)

```

By default, the saptial-auto-correlation scores are only calculated for the variable genes in the Seurat object, here we have 3000. Among the top most variable features in our Seurat object, we find genes coding for hemoglobin; "Hbb-bs" "Hba-a1" "Hba-a2". These are typically expressed in blood vessels which are more randomly distributed across the tissue compared to larger tissue structures. Knowing the spatial auto-correlation can therefore be useful to distinguish genes expressed in larger tissue compartments. One way to make use of this is to restrict the selection of features used for dimensionality reduction and clustering to include ony gene that are highly variable and spatially auto-correlated, and that way avoid clustering based on structures such as blood vessels. 

```{r load_spatgenes, include=FALSE}
load("pre_data/spatgenes")
```
```{r}

head(VariableFeatures(se))
spatgenes[c("Hbb-bs", "Hba-a1", "Hba-a2"), ]

```

Let's plot some of the genes with highest spatial auto-correlation.

```{r print_spatgenes}
head(spatgenes) %>%
  kbl() %>%
  kable_styling()
```
```{r, fig.height=12, fig.width=12, out.width="100%"}

FeatureOverlay(se, features = c("Mbp", "Camk2n1", "Slc6a3", "Th"), 
              sampleids = 1,
              cols = c("lightgray", "mistyrose", "red", "darkred", "black"),
              pt.size = 1.5, 
              add.alpha = TRUE,
              ncol = 2)

```


## Matrix factorization
***

The strength of untargeted whole transcriptome capture is the ability to perform unsupervised analysis and the ability to find spatial gene expression patterns. We've found good use of using non-negative matrix factorization (NNMF or NMF) to find underlying patterns of transcriptomic profiles. This factor analysis, along with various dimensionality reduction techniques, can all be ran via "RunXXX()", where X = the method of choice, e.g.:
<br>
```{r NMF, fig.width=16, fig.height=10, eval = FALSE}

se <- RunNMF(se, nfactors = 40) # Specificy nfactors to choose the number of factors, default=20.

``` 
<br>
While `RunNMF()` is an STutility add-on, others are supported via Seurat (`RunPCA()`, `RunTSNE`, `RunICA()`, `runUMAP()` ) and for all of them, the output are stored in the Seurat object. 

We can then plot a variable number of dimensions across the samples using `ST.DimPlot` or as an overlay using `DimOverlay`. These two functions are similar to the `ST.FeaturePlot` and `FeatureOverlay` but have been adapted to specifically draw dimensionality reduction vectors instead of features.

NOTE: by default, the colorscale of dimensionality reduction vectors will be centered at 0. If we have a dimensionality reduction vector x this means that the range of colors will go from -max(abs(x)) to max(abs(x)). This behaviour is typically desired when plotting e.g. PCA vectors, but for NMF all values are strictly positive so you can disable this centering by setting `center.zero = FALSE`.
<br>
```{r dimplot, fig.height=18, fig.width=12, out.width = "100%"}

cscale <- c("lightgray", "mistyrose", "red", "darkred", "black")

ST.DimPlot(se, 
           dims = 1:10,
           ncol = 2, # Sets the number of columns at dimensions level
           grid.ncol = 2, # Sets the number of columns at sample level
           reduction = "NMF", 
           pt.size = 1, 
           center.zero = F, 
           cols = cscale, 
           show.sb = FALSE)

ST.DimPlot(se, 
           dims = 11:20,
           ncol = 2, 
           grid.ncol = 2, 
           reduction = "NMF", 
           pt.size = 1, 
           center.zero = F, 
           cols = cscale, 
           show.sb = FALSE)


ST.DimPlot(se, 
           dims = 21:30,
           ncol = 2, 
           grid.ncol = 2, 
           reduction = "NMF", 
           pt.size = 1, 
           center.zero = F, 
           cols = cscale, 
           show.sb = FALSE)

ST.DimPlot(se, 
           dims = 31:40,
           ncol = 2, 
           grid.ncol = 2, 
           reduction = "NMF", 
           pt.size = 1, 
           center.zero = F, 
           cols = cscale, 
           show.sb = FALSE)

```
<br>
We can also print a summary of the genes that contribute most to the dimensionality reduction vectors. 

For NMF output which is not centered at 0 looking at the "negative" side of the distribution doesn't really add any valuable information, instead you can aget a barplot summarizing the top most contributing genes using `FactorGeneLoadingPlot`. 
<br>
```{r project_dim}

print(se[["NMF"]])
FactorGeneLoadingPlot(se, factor = 1)

```


## Clustering
***

Clustering is a standard procedure in genomic analysis, and the methods for doing so are numerous. Here we demonstrate an example where we use the result of the factor analysis the previous section. Going through the list of factors (e.g. via `ST:DimPlot(se, dims = [dims you want to look at])`), we can notice dimensions that are "spatially active", i.e. that seems to confer a spatial pattern along their axis. We can extract these dimensions and use as input to e.g. clustering functions. Here, we use all dimensions from the NMF and construct a Shared Nearest Neighbor (SSN) Graph. 
<br>
```{r findneighbours, eval=FALSE}
se <- FindNeighbors(object = se, verbose = FALSE, reduction = "NMF", dims = 1:40)
```
<br>
Followed by clustering using a modularity optimizer
<br>
```{r findclusters, eval=FALSE}
se <- FindClusters(object = se, verbose = FALSE)
```
<br>
And plotting of the clusters spatially
<br>
```{r plot_clusters, fig.height = 5, fig.width = 10, out.width='100%'}

library(RColorBrewer)
n <- 19
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))

ST.FeaturePlot(object = se, features = "seurat_clusters", cols = col_vector, pt.size = 1, ncol = 2)

```
<br>
If you think that the distribution of clusters gets too cluttered, you can also split the view so that only one cluster at the time gets colored, just note that you can only do this for one section at the time (set `ìndex`).
<br>
```{r plot_clusters_split, fig.height=16, fig.width=16, out.width="100%"}

ST.FeaturePlot(object = se, features = "seurat_clusters", pt.size = 1, split.labels = T, indices = 1, show.sb = FALSE, ncol = 5)

ST.FeaturePlot(object = se, features = "seurat_clusters", pt.size = 1, split.labels = T, indices = 2, show.sb = FALSE, ncol = 5)
```
<br>

## Most variable features
***

We can take a specific look at some of the most variable features defined during the normalization step.
<br>
```{r variable_features, fig.height=6, fig.width=12, out.width='100%'}

head(se@assays$SCT@var.features, 20)
top <- se@assays$SCT@var.features

fts <- c("Th", "Mbp", "Nrgn")
for (ftr in fts) {
  p <- FeatureOverlay(se, 
                  features = ftr, 
                  sampleids = 1:2,
                  cols = c("lightgray", "mistyrose", "red", "darkred", "black"),
                  pt.size = 1.5, 
                  pt.alpha = 0.5, 
                 ncols = 2)
  print(p)
}

```
<br>

## Spatial vs. UMAP visualization
***

Another useful feature is that you can now compare the spatial distribution of a gene with the typical "graph embeddings" s.a. UMAP and t-SNE. 
<br>
```{r run_UMAP, eval=FALSE}

# Run UMAP
se <- RunUMAP(se, reduction = "NMF", dims = 1:40, n.neighbors = 10)

````

```{r embedding_vs_ST, fig.width = 12, fig.height = 12}

# Define colors for heatmap
heatmap.colors <- c("lightgray", "mistyrose", "red", "darkred", "black")
fts <- c("Prkcd", "Opalin", "Lamp5")

# plot transformed features expression on UMAP embedding
p.fts <- lapply(fts, function(ftr) {
  FeaturePlot(se, features = ftr, reduction = "umap", order = TRUE, cols = heatmap.colors)
})

# plot transformed features expression on Visium coordinates
p3 <- ST.FeaturePlot(se, features = fts, ncol = 2, grid.ncol = 1, cols = heatmap.colors, pt.size = 1, show.sb = FALSE)

# Construct final plot
cowplot::plot_grid(cowplot::plot_grid(plotlist = p.fts, ncol = 1), p3, ncol = 2, rel_widths = c(1, 1.3))

```

## RGB dimensionality reduction plots
***

One approach to visualize the result of dimensionality reduction is to use the first three dimensions and transform the values into RGB color space. This 3 dimensional space can then be utilized for spatial visualization.
We demonstrate the technique with UMAP, using our factors as input:
<br>
```{r UMAP, val=FALSE, eval=FALSE, warning=FALSE}

se <- RunUMAP(object = se, dims = 1:40, verbose = FALSE, n.components = 3, reduction = "NMF", reduction.name = "umap.3d")

```
<br>
We use the first three dimensions for plotting:
<br>
```{r UMAP_blend, fig.height = 7, fig.width = 12, out.width="100%"}

ST.DimPlot(object = se, dims = 1:3, reduction = "umap.3d", blend = T, pt.size = 1.8)

```

<br>

## DEA
***

Lets try this out by an example. Looking at \ref{plot_clusters_split}, lets say we are interested in cluster 19, and want to answer - "Which genes are significantly up-regulated in this region compared to the overall expression across the tissue?"

```{r de_analysis}

markers <- FindMarkers(se, ident.1 = "19")
head(markers) %>%
  kbl() %>%
  kable_styling()

```

Note that the clusters were already set as the Seurat objects levels. Type `levels(se)` to see the current levels of your object. If other clusters, annotations etc are of interest, set this before by specifying `Idents(se) <- `

Note also, if we are interested in comparing two levels against each other, and not just "one against the rest", we simply add a `ident.2 = ` parameter to the above. 


```{r, fig.height=6, fig.width=12, out.width="100%"}

FeatureOverlay(se, features = "Dsp", 
              sampleids = 1:2,
              cols = c("lightgray", "mistyrose", "red", "darkred", "black"),
              pt.size = 1.5, 
              pt.alpha = 0.5,
              ncol = 2)

```
<br>


&nbsp;
<hr />
<p style="text-align: center;">A work by <a href="mailto: j.bergenstrahle@scilifelab.se">Joseph Bergenstråhle</a> and <a href="mailto: ludvig.larsson@scilifelab.se">Ludvig Larsson</a></p>

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">

<p style="text-align: center;">
    <a href="https://www.spatialresearch.org" class="fa fa-beer"></a>
</p>

&nbsp;
